// vidstack/dist/dev/chunks/chunk-ATST7NI2.js
import { renderRadioGroup } from './chunk-KRJ7Q224.js';
import { ClassManager } from './chunk-XCUHASYJ.js';
import { RadioGroup } from './chunk-MIWMXUN7.js';
import { Radio } from './chunk-P2N5WSLT.js';
import { MenuItems } from './chunk-6F6XCALD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_insert_at_marker_lite, $$_create_template } from 'maverick.js/dom';
import { effect } from 'maverick.js';
import { defineElement } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><span part="label"><!$></span>`);
var AudioMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this._media = useMedia();
  }
  onConnect(el) {
    effect(this._watchControllerDisabled.bind(this));
    effect(this._watchHintText.bind(this));
    const { radioClass, radioCheckClass } = this.$props;
    new ClassManager(el)._observe("media-radio", radioClass)._observe('[part="check"]', radioCheckClass);
  }
  _watchHintText() {
    const { emptyLabel } = this.$props, { audioTrack } = this._media.$store, track = audioTrack();
    this._menu._hint.set(track?.label ?? emptyLabel());
  }
  _watchControllerDisabled() {
    this._menu._disable(this._isDisabled());
  }
  _isDisabled() {
    const { audioTracks } = this._media.$store;
    return audioTracks().length === 0;
  }
  _onChange(event) {
    if (this._isDisabled())
      return;
    const radioGroup = event.target;
    const value = radioGroup.value, index = this._media.audioTracks.toArray().findIndex((track) => track.label.toLowerCase() === value);
    if (index >= 0)
      this._media.remote.changeAudioTrack(index, event);
  }
  _getValue() {
    const { audioTrack } = this._media.$store;
    const track = audioTrack();
    return track ? track.label.toLowerCase() : "";
  }
  _getOptions() {
    const { audioTracks } = this._media.$store;
    return audioTracks().map((track) => ({
      value: track.label.toLowerCase(),
      content: () => (() => {
        const [$$_root, $$_walker] = $$_create_walker($$_templ), $$_expr = $$_walker.nextNode();
        $$_insert_at_marker_lite($$_expr, track.label);
        return $$_root;
      })()
    }));
  }
  render() {
    const { radioGroupClass } = this.$props;
    return renderRadioGroup({
      value: this._getValue.bind(this),
      options: this._getOptions.bind(this),
      radioGroupClass,
      onChange: this._onChange.bind(this)
    });
  }
};
AudioMenuItems.el = defineElement({
  tagName: "media-audio-menu-items",
  props: {
    emptyLabel: "Default",
    radioClass: null,
    radioGroupClass: null,
    radioCheckClass: null
  }
});
AudioMenuItems.register = [RadioGroup, Radio];

export { AudioMenuItems };
