// vidstack/dist/dev/chunks/chunk-MX7CJR52.js
import { menuContext } from './chunk-HSL7QSWM.js';
import { TooltipController } from './chunk-YQQPXCW6.js';
import { FocusVisibleController } from './chunk-IQ5YT2BQ.js';
import { useContext, effect } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';

var MenuButton = class extends Component {
  constructor(instance) {
    super(instance);
    this._menu = useContext(menuContext);
    new FocusVisibleController(instance);
    new TooltipController(instance);
  }
  onAttach(el) {
    this._menu._attachMenuButton(el);
    effect(this._watchDisabled.bind(this));
  }
  onConnect(el) {
    const hint = Array.from(el.querySelectorAll('[slot="hint"]')).pop();
    if (hint) {
      effect(() => {
        const text = this._menu._hint();
        if (text)
          hint.textContent = text;
      });
    }
  }
  _watchDisabled() {
    this._menu._disableMenuButton(this.$props.disabled());
  }
};
MenuButton.el = defineElement({
  tagName: "media-menu-button",
  props: { disabled: false }
});

export { MenuButton };
