// vidstack/dist/dev/chunks/chunk-57KDYM5F.js
import { round } from './chunk-CVLY5S52.js';
import { renderRadioGroup } from './chunk-KRJ7Q224.js';
import { ClassManager } from './chunk-XCUHASYJ.js';
import { RadioGroup } from './chunk-MIWMXUN7.js';
import { Radio } from './chunk-P2N5WSLT.js';
import { MenuItems } from './chunk-6F6XCALD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_insert_at_marker_lite, $$_computed, $$_create_template } from 'maverick.js/dom';
import { computed, effect, peek } from 'maverick.js';
import { defineElement } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><span><!$></span>`);
var $$_templ_2 = /* @__PURE__ */ $$_create_template(`<!$><span part="label"><!$></span>`);
var $$_templ_3 = /* @__PURE__ */ $$_create_template(`<!$><span part="info"><!$></span>`);
var QualityMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this._sortedQualities = computed(() => {
      const { qualities } = this._media.$store;
      return [...qualities()].sort(
        (a, b) => b.height === a.height ? b.bitrate - a.bitrate : b.height - a.height
      );
    });
    this._media = useMedia();
  }
  onConnect(el) {
    effect(this._watchControllerDisabled.bind(this));
    effect(this._watchHintText.bind(this));
    const { radioClass, radioCheckClass } = this.$props;
    new ClassManager(el)._observe("media-radio", radioClass)._observe('[part="check"]', radioCheckClass);
  }
  _watchHintText() {
    const { autoLabel } = this.$props, { autoQuality, quality } = this._media.$store, qualityText = quality() ? quality().height + "p" : "";
    this._menu._hint.set(!autoQuality() ? qualityText : autoLabel() + ` (${qualityText})`);
  }
  _watchControllerDisabled() {
    const { qualities } = this._media.$store;
    this._menu._disable(qualities().length === 0);
  }
  _isDisabled() {
    const { canSetQuality, qualities } = this._media.$store;
    return !canSetQuality() || qualities().length === 0;
  }
  _onChange(event) {
    if (this._isDisabled())
      return;
    const radioGroup = event.target, value = radioGroup.value;
    if (value === "auto") {
      this._media.remote.changeQuality(-1, event);
      return;
    }
    const { qualities } = this._media.$store, index = peek(qualities).findIndex((quality) => this._getQualityId(quality) === value);
    if (index >= 0)
      this._media.remote.changeQuality(index, event);
  }
  _getValue() {
    const { quality, autoQuality } = this._media.$store;
    if (autoQuality())
      return "auto";
    const currentQuality = quality();
    return currentQuality ? this._getQualityId(currentQuality) : "auto";
  }
  _getQualityId(quality) {
    return quality.height + "_" + quality.bitrate;
  }
  _getOptions() {
    const { autoLabel, hideBitrate } = this.$props;
    return [
      { value: "auto", content: () => (() => {
        const [$$_root, $$_walker] = $$_create_walker($$_templ), $$_expr = $$_walker.nextNode();
        $$_insert_at_marker_lite($$_expr, autoLabel);
        return $$_root;
      })() },
      ...this._sortedQualities().map((quality) => {
        const rate = `${round(quality.bitrate / 1e6, 2)} Mbps`;
        return {
          value: this._getQualityId(quality),
          content: () => [
            (() => {
              const [$$_root, $$_walker] = $$_create_walker($$_templ_2), $$_expr = $$_walker.nextNode();
              $$_insert_at_marker_lite($$_expr, quality.height + "p");
              return $$_root;
            })(),
            (() => {
              const $$_signal = $$_computed(
                () => !hideBitrate() && (() => {
                  const [$$_root, $$_walker] = $$_create_walker($$_templ_3), $$_expr = $$_walker.nextNode();
                  $$_insert_at_marker_lite($$_expr, rate);
                  return $$_root;
                })()
              );
              $$_signal();
              return $$_signal;
            })()
          ]
        };
      })
    ];
  }
  render() {
    const { radioGroupClass } = this.$props;
    return renderRadioGroup({
      value: this._getValue.bind(this),
      options: this._getOptions.bind(this),
      radioGroupClass,
      onChange: this._onChange.bind(this)
    });
  }
};
QualityMenuItems.el = defineElement({
  tagName: "media-quality-menu-items",
  props: {
    autoLabel: "Auto",
    hideBitrate: false,
    radioGroupClass: null,
    radioClass: null,
    radioCheckClass: null
  }
});
QualityMenuItems.register = [RadioGroup, Radio];

export { QualityMenuItems };
