// vidstack/dist/dev/chunks/chunk-QVMUDSL4.js
import { toggleButtonProps, ToggleButton } from './chunk-252ZP55V.js';
import { setARIALabel } from './chunk-HQQTSFXD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-OSS74BPE.js';
import { replay_default } from './chunk-NRSSXRVS.js';
import { play_default } from './chunk-N7ILVN7U.js';
import { pause_default } from './chunk-2DGYZ3AT.js';
import { $$_create_component } from 'maverick.js/dom';
import { defineElement } from 'maverick.js/element';

var PlayButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this._keyShortcut = "togglePaused";
  }
  onAttach(el) {
    this._media = useMedia();
    this._pressed = this._isPressed.bind(this);
    super.onAttach(el);
    setARIALabel(el, this._getLabel.bind(this));
    const { paused, ended } = this._media.$store;
    this.setAttributes({
      "data-paused": paused,
      "data-ended": ended
    });
  }
  _onPress(event) {
    const remote = this._media.remote;
    this._pressed() ? remote.pause(event) : remote.play(event);
  }
  _isPressed() {
    const { paused } = this._media.$store;
    return !paused();
  }
  _getLabel() {
    const { paused } = this._media.$store;
    return paused() ? "Play" : "Pause";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: play_default, slot: "play" }),
      $$_create_component(Icon, { paths: replay_default, slot: "replay" }),
      $$_create_component(Icon, { paths: pause_default, slot: "pause" })
    ];
  }
};
PlayButton.el = defineElement({
  tagName: "media-play-button",
  props: toggleButtonProps
});

export { PlayButton };
