// vidstack@0.6.9/dist/dev/chunks/chunk-37JVNHU6.js
import { Tooltip } from './chunk-DON7JMVQ.js';
import { VolumeSlider } from './chunk-JBX3YXBH.js';
import { SliderThumbnail } from './chunk-MJRWGNJB.js';
import { SliderValue } from './chunk-UKG5Z2SB.js';
import { SliderVideo } from './chunk-AWEYXSJQ.js';
import { TimeSlider } from './chunk-5PVPU33U.js';
import { Slider } from './chunk-LWTGALKS.js';
import { Time } from './chunk-LZTNZZHJ.js';
import { PlaybackRateMenuItems } from './chunk-BLODLME6.js';
import { Player } from './chunk-VI4ON6IC.js';
import { Poster } from './chunk-VTFMYSE2.js';
import { QualityMenuButton } from './chunk-SJD42WNH.js';
import { QualityMenuItems } from './chunk-57KDYM5F.js';
import { SeekButton } from './chunk-QTDX7HDE.js';
import { Menu } from './chunk-2NMX37TU.js';
import { MuteButton } from './chunk-XBG5E767.js';
import { Outlet } from './chunk-FY36WF27.js';
import { PIPButton } from './chunk-2DWMEI5Q.js';
import { PlayButton } from './chunk-QVMUDSL4.js';
import { PlaybackRateMenuButton } from './chunk-UOQZCGPO.js';
import { CaptionsMenuItems } from './chunk-E5ZWM3RR.js';
import { Captions } from './chunk-L4H3J6NX.js';
import { ChaptersMenuItems } from './chunk-EYRSCQBE.js';
import { Thumbnail } from './chunk-AJO5IWRQ.js';
import { FullscreenButton } from './chunk-V32T6ASS.js';
import { Gesture } from './chunk-WIWKZOJA.js';
import { LiveIndicator } from './chunk-LTLQUHKO.js';
import { AudioMenuButton } from './chunk-MQYRFZNM.js';
import { AudioMenuItems } from './chunk-ATST7NI2.js';
import { RadioGroup } from './chunk-MIWMXUN7.js';
import { Radio } from './chunk-P2N5WSLT.js';
import { MenuItems } from './chunk-6F6XCALD.js';
import { BufferingIndicator } from './chunk-LFUONOET.js';
import { CaptionButton } from './chunk-JCI4LGRO.js';
import { ToggleButton } from './chunk-252ZP55V.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { CaptionsMenuButton } from './chunk-Y76E56TL.js';
import { MenuButton } from './chunk-MX7CJR52.js';
import { Icon } from './chunk-OSS74BPE.js';
import { settings_default } from './chunk-764WJ7YC.js';
import { seek_forward_10_default } from './chunk-B37QG5XE.js';
import { seek_backward_10_default } from './chunk-FNH5S7FD.js';
import { chapters_default } from './chunk-FPO3FMZ7.js';
import { defineElement, Component, registerLiteCustomElement } from 'maverick.js/element';
import { createContext, provideContext, computed, signal, effect, useContext } from 'maverick.js';
import { $$_clone, $$_setup_custom_element, $$_insert_lite, $$_create_component, $$_attr, $$_scoped, $$_create_template } from 'maverick.js/dom';
import { listenEvent } from 'maverick.js/std';

function i18n(translations, key) {
  return translations()?.[key] ?? key;
}
var communitySkinContext = createContext();
function useCommunitySkin() {
  return useContext(communitySkinContext);
}
function useCommunitySkinI18n() {
  return useCommunitySkin().$props.translations;
}

// src/player/skins/community/shared.tsx
var $$_templ = /* @__PURE__ */ $$_create_template(`<media-play-button default-appearance="" mk-d><media-tooltip mk-d><span slot="play"></span><span slot="pause"></span></media-tooltip></media-play-button>`);
var $$_templ_2 = /* @__PURE__ */ $$_create_template(`<media-mute-button default-appearance="" mk-d><media-tooltip mk-d><span slot="mute"></span><span slot="unmute"></span></media-tooltip></media-mute-button>`);
var $$_templ_3 = /* @__PURE__ */ $$_create_template(`<media-caption-button default-appearance="" mk-d><media-tooltip mk-d><span slot="on"></span><span slot="off"></span></media-tooltip></media-caption-button>`);
var $$_templ_4 = /* @__PURE__ */ $$_create_template(`<media-pip-button default-appearance="" mk-d><media-tooltip mk-d><span slot="enter"></span><span slot="exit"></span></media-tooltip></media-pip-button>`);
var $$_templ_5 = /* @__PURE__ */ $$_create_template(`<media-fullscreen-button default-appearance="" mk-d><media-tooltip position="top right" mk-d><span slot="enter"></span><span slot="exit"></span></media-tooltip></media-fullscreen-button>`);
var $$_templ_6 = /* @__PURE__ */ $$_create_template(`<media-seek-button mk-d><media-tooltip mk-d><span></span></media-tooltip></media-seek-button>`);
var $$_templ_7 = /* @__PURE__ */ $$_create_template(`<media-volume-slider mk-d><media-slider-value type="pointer" format="percent" slot="preview" mk-d></media-slider-value></media-volume-slider>`);
var $$_templ_8 = /* @__PURE__ */ $$_create_template(`<span part="main-title"></span>`);
var $$_templ_9 = $$_templ_8;
var $$_templ_10 = /* @__PURE__ */ $$_create_template(`<media-menu part="chapters-menu" mk-d><media-menu-button mk-d><media-tooltip mk-d></media-tooltip></media-menu-button><media-chapters-menu-items mk-d></media-chapters-menu-items></media-menu>`);
var $$_templ_11 = /* @__PURE__ */ $$_create_template(`<media-menu part="settings-menu" mk-d><media-menu-button mk-d><media-tooltip mk-d></media-tooltip></media-menu-button><media-menu-items mk-d><media-menu mk-d><media-audio-menu-button mk-d></media-audio-menu-button><media-audio-menu-items mk-d></media-audio-menu-items></media-menu><media-menu mk-d><media-playback-rate-menu-button mk-d></media-playback-rate-menu-button><media-playback-rate-menu-items mk-d></media-playback-rate-menu-items></media-menu><media-menu mk-d><media-quality-menu-button mk-d></media-quality-menu-button><media-quality-menu-items mk-d></media-quality-menu-items></media-menu><media-menu mk-d><media-captions-menu-button mk-d></media-captions-menu-button><media-captions-menu-items mk-d></media-captions-menu-items></media-menu></media-menu-items></media-menu>`);
var $$_templ_12 = /* @__PURE__ */ $$_create_template(`<div part="gestures"><media-gesture event="pointerup" action="toggle:paused" mk-d></media-gesture><media-gesture event="pointerup" action="toggle:user-idle" mk-d></media-gesture><media-gesture event="dblpointerup" action="toggle:fullscreen" mk-d></media-gesture><media-gesture event="dblpointerup" action="seek:-10" mk-d></media-gesture><media-gesture event="dblpointerup" action="seek:10" mk-d></media-gesture></div>`);
var $$_templ_13 = /* @__PURE__ */ $$_create_template(`<media-time-slider mk-d><div slot="preview"><media-slider-thumbnail mk-d></media-slider-thumbnail><div part="chapter-title"></div><media-slider-value type="pointer" format="time" mk-d></media-slider-value></div></media-time-slider>`);
var $$_templ_14 = /* @__PURE__ */ $$_create_template(`<div part="time-group"><media-time type="current" mk-d></media-time><div part="time-divider">/</div><media-time type="duration" mk-d></media-time></div>`);
function PlayButton2({
  part,
  tooltip = "top left"
}) {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el_2.nextSibling;
    $$_attr($$_root, "part", part);
    $$_scoped(() => {
      $$_attr($$_el, "position", tooltip);
      $$_scoped(() => {
        $$_insert_lite($$_el_2, () => i18n(lang, "Play"));
        $$_insert_lite($$_el_3, () => i18n(lang, "Pause"));
      }, $$_setup_custom_element($$_el));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function MuteButton2({ tooltip = "top center" }) {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_2), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el_2.nextSibling;
    $$_scoped(() => {
      $$_attr($$_el, "position", tooltip);
      $$_scoped(() => {
        $$_insert_lite($$_el_2, () => i18n(lang, "Mute"));
        $$_insert_lite($$_el_3, () => i18n(lang, "Unmute"));
      }, $$_setup_custom_element($$_el));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function CaptionButton2({ tooltip = "top center" }) {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_3), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el_2.nextSibling;
    $$_scoped(() => {
      $$_attr($$_el, "position", tooltip);
      $$_scoped(() => {
        $$_insert_lite($$_el_2, () => i18n(lang, "Closed-Captions On"));
        $$_insert_lite($$_el_3, () => i18n(lang, "Closed-Captions Off"));
      }, $$_setup_custom_element($$_el));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function PiPButton() {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_4), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el_2.nextSibling;
    $$_scoped(() => {
      $$_scoped(() => {
        $$_insert_lite($$_el_2, () => i18n(lang, "Enter PiP"));
        $$_insert_lite($$_el_3, () => i18n(lang, "Exit PiP"));
      }, $$_setup_custom_element($$_el));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function FullscreenButton2() {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_5), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el_2.nextSibling;
    $$_scoped(() => {
      $$_scoped(() => {
        $$_insert_lite($$_el_2, () => i18n(lang, "Enter Fullscreen"));
        $$_insert_lite($$_el_3, () => i18n(lang, "Exit Fullscreen"));
      }, $$_setup_custom_element($$_el));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function SeekButton2({
  seconds,
  tooltip = "top center"
}) {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_6), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild;
    $$_attr($$_root, "seconds", seconds);
    $$_scoped(() => {
      $$_insert_lite(
        $$_root,
        $$_create_component(Icon, { paths: seconds >= 0 ? seek_forward_10_default : seek_backward_10_default }),
        $$_el
      );
      $$_attr($$_el, "position", tooltip);
      $$_scoped(() => {
        $$_insert_lite($$_el_2, () => i18n(lang, seconds >= 0 ? "Seek Forward" : "Seek Backward"));
      }, $$_setup_custom_element($$_el));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function VolumeSlider2() {
  return (() => {
    const $$_root = $$_clone($$_templ_7), $$_el = $$_root.firstChild;
    $$_scoped(() => {
      $$_setup_custom_element($$_el);
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function ChapterTitleOrMainTitle() {
  const {
    $media: { title, textTracks, started }
  } = useCommunitySkin();
  const chapterTitle = signal(""), mainTitle = computed(() => started() ? chapterTitle() || title() : title());
  effect(() => {
    const track = textTracks().find(
      (track2) => track2.kind === "chapters" && track2.mode === "showing"
    );
    track && listenEvent(track, "cue-change", () => {
      chapterTitle.set(track.activeCues[0]?.text || "");
    });
    return () => chapterTitle.set("");
  });
  return (() => {
    const $$_root = $$_clone($$_templ_9);
    $$_insert_lite($$_root, mainTitle);
    return $$_root;
  })();
}
function ChaptersMenu({
  position = "bottom",
  tooltip = "bottom center"
}) {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_10), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el.nextSibling;
    $$_attr($$_root, "position", position);
    $$_scoped(() => {
      $$_scoped(() => {
        $$_insert_lite($$_el, $$_create_component(Icon, { paths: chapters_default }), $$_el_2);
        $$_attr($$_el_2, "position", tooltip);
        $$_scoped(() => {
          $$_insert_lite($$_el_2, () => i18n(lang, "Chapters"));
        }, $$_setup_custom_element($$_el_2));
      }, $$_setup_custom_element($$_el));
      $$_setup_custom_element($$_el_3);
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function SettingsMenu({
  position = "bottom",
  tooltip = "bottom right"
}) {
  const lang = useCommunitySkinI18n();
  return (() => {
    const $$_root = $$_clone($$_templ_11), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el.nextSibling, $$_el_4 = $$_el_3.firstChild, $$_el_5 = $$_el_4.firstChild, $$_el_6 = $$_el_5.nextSibling, $$_el_7 = $$_el_4.nextSibling, $$_el_8 = $$_el_7.firstChild, $$_el_9 = $$_el_8.nextSibling, $$_el_10 = $$_el_7.nextSibling, $$_el_11 = $$_el_10.firstChild, $$_el_12 = $$_el_11.nextSibling, $$_el_13 = $$_el_10.nextSibling, $$_el_14 = $$_el_13.firstChild, $$_el_15 = $$_el_14.nextSibling;
    $$_attr($$_root, "position", position);
    $$_scoped(() => {
      $$_scoped(() => {
        $$_insert_lite($$_el, $$_create_component(Icon, { paths: settings_default, rotate: true }), $$_el_2);
        $$_attr($$_el_2, "position", tooltip);
        $$_scoped(() => {
          $$_insert_lite($$_el_2, () => i18n(lang, "Settings"));
        }, $$_setup_custom_element($$_el_2));
      }, $$_setup_custom_element($$_el));
      $$_scoped(() => {
        $$_scoped(() => {
          $$_setup_custom_element($$_el_5, { label: () => i18n(lang, "Audio") });
          $$_setup_custom_element($$_el_6, { emptyLabel: () => i18n(lang, "Default") });
        }, $$_setup_custom_element($$_el_4));
        $$_scoped(() => {
          $$_setup_custom_element($$_el_8, { label: () => i18n(lang, "Speed") });
          $$_setup_custom_element($$_el_9, { normalLabel: () => i18n(lang, "Normal") });
        }, $$_setup_custom_element($$_el_7));
        $$_scoped(() => {
          $$_setup_custom_element($$_el_11, { label: () => i18n(lang, "Quality") });
          $$_setup_custom_element($$_el_12, { autoLabel: () => i18n(lang, "Auto") });
        }, $$_setup_custom_element($$_el_10));
        $$_scoped(() => {
          $$_setup_custom_element($$_el_14, { label: () => i18n(lang, "Captions") });
          $$_setup_custom_element($$_el_15, { offLabel: () => i18n(lang, "Off") });
        }, $$_setup_custom_element($$_el_13));
      }, $$_setup_custom_element($$_el_3));
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function VideoGestures() {
  return (() => {
    const $$_root = $$_clone($$_templ_12), $$_el = $$_root.firstChild, $$_el_2 = $$_el.nextSibling, $$_el_3 = $$_el_2.nextSibling, $$_el_4 = $$_el_3.nextSibling, $$_el_5 = $$_el_4.nextSibling;
    $$_setup_custom_element($$_el);
    $$_setup_custom_element($$_el_2);
    $$_setup_custom_element($$_el_3);
    $$_setup_custom_element($$_el_4);
    $$_setup_custom_element($$_el_5);
    return $$_root;
  })();
}
function TimeSlider2() {
  return (() => {
    const $$_root = $$_clone($$_templ_13), $$_el = $$_root.firstChild, $$_el_2 = $$_el.firstChild, $$_el_3 = $$_el_2.nextSibling, $$_el_4 = $$_el_3.nextSibling;
    $$_scoped(() => {
      $$_setup_custom_element($$_el_2);
      $$_setup_custom_element($$_el_4);
    }, $$_setup_custom_element($$_root));
    return $$_root;
  })();
}
function TimeGroup() {
  return (() => {
    const $$_root = $$_clone($$_templ_14), $$_el = $$_root.firstChild, $$_el_2 = $$_el.nextSibling, $$_el_3 = $$_el_2.nextSibling;
    $$_setup_custom_element($$_el);
    $$_setup_custom_element($$_el_3);
    return $$_root;
  })();
}

// src/player/skins/community/audio.tsx
var $$_templ2 = /* @__PURE__ */ $$_create_template(`<div part="media-ui aod live:dvr mobile"><media-captions mk-d></media-captions><div part="controls"><div part="controls-group"></div><div part="controls-group"></div><div part="controls-group"><media-time type="current" mk-d></media-time><div part="controls-spacer"></div><media-time type="duration" mk-d></media-time></div><div part="controls-group"><div part="controls-spacer"></div><div part="controls-spacer"></div></div></div></div>`);
var $$_templ_22 = /* @__PURE__ */ $$_create_template(`<div part="media-ui aod live:dvr desktop"><media-captions mk-d></media-captions><div part="controls"><div part="controls-group"></div><div part="controls-group"></div></div></div>`);
function renderAudio(isMobile) {
  return isMobile ? MobileUI() : DesktopUI();
}
function MobileUI() {
  return (() => {
    const $$_root = $$_clone($$_templ2), $$_el = $$_root.firstChild, $$_el_2 = $$_el.nextSibling, $$_el_3 = $$_el_2.firstChild, $$_el_4 = $$_el_3.nextSibling, $$_el_5 = $$_el_4.nextSibling, $$_el_6 = $$_el_5.firstChild, $$_el_7 = $$_el_6.nextSibling, $$_el_8 = $$_el_7.nextSibling, $$_el_9 = $$_el_5.nextSibling, $$_el_10 = $$_el_9.firstChild, $$_el_11 = $$_el_10.nextSibling;
    $$_setup_custom_element($$_el);
    $$_insert_lite($$_el_3, $$_create_component(MuteButton2, { tooltip: "top left" }), null);
    $$_insert_lite($$_el_3, $$_create_component(ChapterTitleOrMainTitle), null);
    $$_insert_lite($$_el_3, $$_create_component(CaptionButton2), null);
    $$_insert_lite($$_el_3, $$_create_component(ChaptersMenu, { tooltip: "top center", position: "top" }), null);
    $$_insert_lite($$_el_3, $$_create_component(SettingsMenu, { tooltip: "top right", position: "top" }), null);
    $$_insert_lite($$_el_4, $$_create_component(TimeSlider2));
    $$_setup_custom_element($$_el_6);
    $$_setup_custom_element($$_el_8);
    $$_insert_lite($$_el_9, $$_create_component(SeekButton2, { seconds: -10, tooltip: "top left" }), $$_el_11);
    $$_insert_lite($$_el_9, $$_create_component(PlayButton2, { tooltip: "top center" }), $$_el_11);
    $$_insert_lite($$_el_9, $$_create_component(SeekButton2, { seconds: 10 }), $$_el_11);
    return $$_root;
  })();
}
function DesktopUI() {
  return (() => {
    const $$_root = $$_clone($$_templ_22), $$_el = $$_root.firstChild, $$_el_2 = $$_el.nextSibling, $$_el_3 = $$_el_2.firstChild, $$_el_4 = $$_el_3.nextSibling;
    $$_setup_custom_element($$_el);
    $$_insert_lite($$_el_3, $$_create_component(TimeSlider2));
    $$_insert_lite($$_el_4, $$_create_component(SeekButton2, { seconds: -10, tooltip: "top left" }), null);
    $$_insert_lite($$_el_4, $$_create_component(PlayButton2, { tooltip: "top center" }), null);
    $$_insert_lite($$_el_4, $$_create_component(SeekButton2, { seconds: 10 }), null);
    $$_insert_lite($$_el_4, $$_create_component(TimeGroup), null);
    $$_insert_lite($$_el_4, $$_create_component(ChapterTitleOrMainTitle), null);
    $$_insert_lite($$_el_4, $$_create_component(MuteButton2, { tooltip: "top center" }), null);
    $$_insert_lite($$_el_4, $$_create_component(VolumeSlider2), null);
    $$_insert_lite($$_el_4, $$_create_component(CaptionButton2), null);
    $$_insert_lite($$_el_4, $$_create_component(ChaptersMenu, { tooltip: "top center", position: "top" }), null);
    $$_insert_lite($$_el_4, $$_create_component(SettingsMenu, { tooltip: "top right", position: "top" }), null);
    return $$_root;
  })();
}
var $$_templ3 = /* @__PURE__ */ $$_create_template(`<div part="media-ui vod live:dvr mobile"><media-captions mk-d></media-captions><media-buffering-indicator mk-d></media-buffering-indicator><div part="scrim"></div><div part="controls"><div part="controls-group"><div part="controls-spacer"></div></div><div part="controls-group"></div><div part="controls-group"><div part="controls-spacer"></div></div><div part="controls-group"></div></div><div part="start-duration"><media-time type="duration" mk-d></media-time></div></div>`);
var $$_templ_23 = /* @__PURE__ */ $$_create_template(`<div part="media-ui vod live:dvr desktop "><media-captions mk-d></media-captions><media-buffering-indicator mk-d></media-buffering-indicator><div part="scrim"></div><div part="controls"><div part="controls-group"><div part="controls-spacer"></div></div><div part="controls-group"></div><div part="controls-group"></div><div part="controls-group"></div></div></div>`);
function renderVideo(isMobile) {
  return isMobile ? MobileUI2() : DesktopUI2();
}
function MobileUI2() {
  return (() => {
    const $$_root = $$_clone($$_templ3), $$_el = $$_root.firstChild, $$_el_2 = $$_el.nextSibling, $$_el_3 = $$_el_2.nextSibling, $$_el_4 = $$_el_3.nextSibling, $$_el_5 = $$_el_4.firstChild; $$_el_5.firstChild; const $$_el_7 = $$_el_5.nextSibling, $$_el_8 = $$_el_7.nextSibling, $$_el_9 = $$_el_8.firstChild, $$_el_10 = $$_el_8.nextSibling, $$_el_11 = $$_el_4.nextSibling, $$_el_12 = $$_el_11.firstChild;
    $$_insert_lite($$_root, $$_create_component(VideoGestures), $$_el);
    $$_setup_custom_element($$_el);
    $$_setup_custom_element($$_el_2);
    $$_insert_lite($$_el_5, $$_create_component(CaptionButton2, { tooltip: "bottom center" }), null);
    $$_insert_lite($$_el_5, $$_create_component(ChaptersMenu), null);
    $$_insert_lite($$_el_5, $$_create_component(SettingsMenu, { tooltip: "bottom center" }), null);
    $$_insert_lite($$_el_5, $$_create_component(MuteButton2, { tooltip: "bottom right" }), null);
    $$_insert_lite($$_el_7, $$_create_component(PlayButton2, { tooltip: "top center" }));
    $$_insert_lite($$_el_8, $$_create_component(TimeGroup), $$_el_9);
    $$_insert_lite($$_el_8, $$_create_component(ChapterTitleOrMainTitle), $$_el_9);
    $$_insert_lite($$_el_8, $$_create_component(FullscreenButton2), null);
    $$_insert_lite($$_el_10, $$_create_component(TimeSlider2));
    $$_setup_custom_element($$_el_12);
    return $$_root;
  })();
}
function DesktopUI2() {
  return (() => {
    const $$_root = $$_clone($$_templ_23), $$_el = $$_root.firstChild, $$_el_2 = $$_el.nextSibling, $$_el_3 = $$_el_2.nextSibling, $$_el_4 = $$_el_3.nextSibling, $$_el_5 = $$_el_4.firstChild; $$_el_5.firstChild; const $$_el_7 = $$_el_5.nextSibling, $$_el_8 = $$_el_7.nextSibling, $$_el_9 = $$_el_8.nextSibling;
    $$_insert_lite($$_root, $$_create_component(VideoGestures), $$_el);
    $$_setup_custom_element($$_el);
    $$_setup_custom_element($$_el_2);
    $$_insert_lite($$_el_5, $$_create_component(ChaptersMenu), null);
    $$_insert_lite($$_el_5, $$_create_component(SettingsMenu), null);
    $$_insert_lite($$_el_8, $$_create_component(TimeSlider2));
    $$_insert_lite($$_el_9, $$_create_component(PlayButton2), null);
    $$_insert_lite($$_el_9, $$_create_component(MuteButton2), null);
    $$_insert_lite($$_el_9, $$_create_component(VolumeSlider2), null);
    $$_insert_lite($$_el_9, $$_create_component(TimeGroup), null);
    $$_insert_lite($$_el_9, $$_create_component(ChapterTitleOrMainTitle), null);
    $$_insert_lite($$_el_9, $$_create_component(CaptionButton2), null);
    $$_insert_lite($$_el_9, $$_create_component(PiPButton), null);
    $$_insert_lite($$_el_9, $$_create_component(FullscreenButton2), null);
    return $$_root;
  })();
}

// src/player/skins/community/skin.tsx
var CommunitySkin = class extends Component {
  constructor(instance) {
    super(instance);
    this._media = useMedia();
    provideContext(communitySkinContext, {
      $props: this.$props,
      $media: this._media.$store
    });
  }
  /** We need this to compute and save the layout type to prevent unnecessary re-renders. */
  _getLayoutType() {
    const { viewType, streamType } = this._media.$store;
    return viewType() === "audio" ? streamType().includes("live") ? "audio:live" : "audio" : streamType().endsWith("live") ? "video:live" : "video";
  }
  onAttach() {
    this.setAttributes({
      "data-audio": this._isAudio.bind(this),
      "data-video": this._isVideo.bind(this),
      "data-mobile": this._isMobile.bind(this)
    });
  }
  _isAudio() {
    const { viewType } = this._media.$store;
    return viewType() === "audio";
  }
  _isVideo() {
    const { viewType } = this._media.$store;
    return viewType() !== "audio";
  }
  _isMobile() {
    const { breakpointX } = this._media.$store;
    return breakpointX() === "sm";
  }
  render() {
    const $layoutType = computed(this._getLayoutType.bind(this)), $isMobile = computed(this._isMobile.bind(this));
    return () => {
      const render = $layoutType().startsWith("video") ? renderVideo : renderAudio;
      return render($isMobile());
    };
  }
};
CommunitySkin.el = defineElement({
  tagName: "media-community-skin",
  nohydrate: true,
  props: {
    translations: null
  }
});
CommunitySkin.register = getUIComponents;

// src/register.ts
function getUIComponents() {
  return [
    Poster,
    BufferingIndicator,
    Tooltip,
    Thumbnail,
    PlayButton,
    MuteButton,
    PIPButton,
    FullscreenButton,
    CaptionButton,
    SeekButton,
    Slider,
    TimeSlider,
    VolumeSlider,
    SliderThumbnail,
    SliderValue,
    SliderVideo,
    Time,
    ToggleButton,
    LiveIndicator,
    Captions,
    Gesture,
    Menu,
    MenuButton,
    MenuItems,
    RadioGroup,
    Radio,
    AudioMenuButton,
    AudioMenuItems,
    CaptionsMenuButton,
    CaptionsMenuItems,
    ChaptersMenuItems,
    QualityMenuButton,
    QualityMenuItems,
    PlaybackRateMenuButton,
    PlaybackRateMenuItems
  ];
}
function registerAllElements() {
  [Player, Outlet, ...getUIComponents(), CommunitySkin].map(registerLiteCustomElement);
}

export { CommunitySkin, getUIComponents, registerAllElements };
