// vidstack@0.6.9/dist/server/index.js
export { CommunitySkin } from './chunks/chunk-7EYMAZ32.js';
export { Tooltip } from './chunks/chunk-KKO6J3BU.js';
export { VolumeSlider } from './chunks/chunk-OTIMLXP6.js';
export { SliderThumbnail } from './chunks/chunk-B5WOLIJG.js';
export { SliderValue } from './chunks/chunk-3MX4LIKX.js';
export { SliderVideo } from './chunks/chunk-2XBVKXSY.js';
export { TimeSlider } from './chunks/chunk-FCHXUPLA.js';
export { Slider, sliderProps } from './chunks/chunk-PAFQPZS6.js';
import './chunks/chunk-U43VNZZW.js';
export { SliderStoreFactory } from './chunks/chunk-53B74NMI.js';
export { Time } from './chunks/chunk-5FE2F5AG.js';
export { PlaybackRateMenuItems } from './chunks/chunk-ZXAPY2BF.js';
export { AudioTrackList, FullscreenController, List, MEDIA_KEY_SHORTCUTS, MediaRemoteControl, MediaStoreFactory, MediaUserController, Player, ScreenOrientationController, TextRenderers, TextTrackList, TimeRange, VideoQualityList, canFullscreen, getTimeRangesEnd, getTimeRangesStart, isAudioProvider, isHLSProvider, isHTMLAudioElement, isHTMLMediaElement, isHTMLVideoElement, isVideoProvider, softResetMediaStore } from './chunks/chunk-HGP4J5WC.js';
import './chunks/chunk-H3U4YXTB.js';
import './chunks/chunk-2CNJRAGE.js';
import './chunks/chunk-W4CS7QKB.js';
import './chunks/chunk-TTEXQSM2.js';
import './chunks/chunk-JELAJF2G.js';
export { Poster } from './chunks/chunk-2QEGRBFM.js';
export { QualityMenuButton } from './chunks/chunk-57OL5R3S.js';
export { QualityMenuItems } from './chunks/chunk-A4DBFVNS.js';
export { SeekButton } from './chunks/chunk-MY2FGNE2.js';
export { Menu } from './chunks/chunk-PALBLMCC.js';
export { MuteButton } from './chunks/chunk-BQKGS4S5.js';
export { Outlet } from './chunks/chunk-NKBW4B6B.js';
import './chunks/chunk-TEKTM4DZ.js';
import './chunks/chunk-M727CDUQ.js';
export { PIPButton } from './chunks/chunk-3JMJZNWQ.js';
export { PlayButton } from './chunks/chunk-COS3VFYB.js';
export { PlaybackRateMenuButton } from './chunks/chunk-NHIC66MT.js';
export { CaptionsMenuItems } from './chunks/chunk-JSSTSFA2.js';
export { Captions } from './chunks/chunk-R43I4DRV.js';
export { ChaptersMenuItems } from './chunks/chunk-4BWBKFM6.js';
export { Thumbnail } from './chunks/chunk-MCANHAYI.js';
import './chunks/chunk-JY53PGKC.js';
import './chunks/chunk-N2X5VJTG.js';
export { FullscreenButton } from './chunks/chunk-MBBDB3JG.js';
export { Gesture } from './chunks/chunk-ILGPP2W6.js';
export { LiveIndicator } from './chunks/chunk-J3NSFYPM.js';
export { AudioMenuButton } from './chunks/chunk-F5YZFZEC.js';
export { AudioMenuItems } from './chunks/chunk-4F5G36NV.js';
import './chunks/chunk-TE367BQX.js';
import './chunks/chunk-XCUHASYJ.js';
export { RadioGroup } from './chunks/chunk-5RWMH5KW.js';
export { Radio } from './chunks/chunk-SKUE5DXF.js';
import './chunks/chunk-4L7R5OAB.js';
export { MenuItems } from './chunks/chunk-3QXOFPLA.js';
export { BufferingIndicator } from './chunks/chunk-X2EJI2SD.js';
export { CaptionButton } from './chunks/chunk-ZZBQCJ7P.js';
import { TEXT_TRACK_READY_STATE } from './chunks/chunk-ZG64VF3W.js';
export { TextTrack, isTrackCaptionKind } from './chunks/chunk-ZG64VF3W.js';
import './chunks/chunk-72VYJ2OF.js';
import './chunks/chunk-NOPW2CG3.js';
export { ToggleButton, toggleButtonProps } from './chunks/chunk-RGSXXUQI.js';
import './chunks/chunk-EVP5E6AO.js';
import './chunks/chunk-JTTKMFDW.js';
export { mediaContext } from './chunks/chunk-ZIBTY3GT.js';
export { CaptionsMenuButton } from './chunks/chunk-JTT52VRX.js';
import './chunks/chunk-WJNOKALV.js';
export { MenuButton } from './chunks/chunk-BH4KDLCZ.js';
import './chunks/chunk-HSL7QSWM.js';
import './chunks/chunk-HPMGEKCS.js';
import './chunks/chunk-YQQPXCW6.js';
import './chunks/chunk-RBTESKWW.js';
import './chunks/chunk-H53PUAI2.js';
import './chunks/chunk-AF5JW3F2.js';
import './chunks/chunk-D3BAJ33A.js';
import './chunks/chunk-64OI64HO.js';
import './chunks/chunk-VE2HQ3IX.js';
import './chunks/chunk-G2UYOOKJ.js';
import './chunks/chunk-WEY2CJH7.js';
import './chunks/chunk-FEPBIWLU.js';
import './chunks/chunk-44XSWTBR.js';
import './chunks/chunk-VOO5DFHH.js';
import './chunks/chunk-NH3HUDRC.js';
import './chunks/chunk-RD4ZCUPM.js';
import './chunks/chunk-54OFHFXV.js';
import './chunks/chunk-GU4UUVUD.js';
import './chunks/chunk-4EVYWGVN.js';
import './chunks/chunk-ZIESC7MH.js';
import './chunks/chunk-3D3GWOB4.js';
import './chunks/chunk-NTQ54KBX.js';
import './chunks/chunk-ATG6EPJK.js';
import './chunks/chunk-DOCK5KNM.js';
import './chunks/chunk-DB7MROLC.js';
import './chunks/chunk-LJYBPGTM.js';
import './chunks/chunk-LO3KCXEW.js';
import './chunks/chunk-KM3IPD4E.js';
import './chunks/chunk-5XOURGNQ.js';
import './chunks/chunk-62ZCYUNF.js';
import './chunks/chunk-2LQRO2BX.js';
import './chunks/chunk-ZLWKC5UX.js';
import './chunks/chunk-MUUSA6L7.js';
import './chunks/chunk-66N63LA6.js';
import './chunks/chunk-FVWDCVES.js';
import './chunks/chunk-3Q6XOPNZ.js';
import './chunks/chunk-AV3YVPW4.js';
import './chunks/chunk-GWZRNBF5.js';
import './chunks/chunk-AL6CJSGK.js';
import './chunks/chunk-TSIGXURK.js';
import './chunks/chunk-QNSYUQKU.js';
import './chunks/chunk-IAD3FOJO.js';
import './chunks/chunk-MU7S3WSI.js';
import './chunks/chunk-GCPAVK2F.js';
import './chunks/chunk-SZEXS3HA.js';
import './chunks/chunk-OQFUC4AZ.js';
import './chunks/chunk-6BHOVF5C.js';
import './chunks/chunk-5XEU3B5K.js';
import './chunks/chunk-KSBA7JS7.js';
import './chunks/chunk-J4ZDSS5K.js';
import './chunks/chunk-NXMYBTP3.js';
import './chunks/chunk-HZUXRM2O.js';
import './chunks/chunk-MNIV2J3A.js';
import './chunks/chunk-HBJMUKLR.js';
import './chunks/chunk-A722DZWS.js';
import './chunks/chunk-OUBJ7HME.js';
import './chunks/chunk-WTETBRVG.js';
import './chunks/chunk-BD76OTCC.js';
import './chunks/chunk-5OZFS7YS.js';
import './chunks/chunk-CJPTKVK4.js';
import './chunks/chunk-GIVKBKHP.js';
import './chunks/chunk-VCERFN4C.js';
import './chunks/chunk-PGA42FHG.js';
import './chunks/chunk-WVCCCJY3.js';
import './chunks/chunk-EQENEGDX.js';
import './chunks/chunk-6GRSLK6Y.js';
import './chunks/chunk-47QOMXRA.js';
import './chunks/chunk-AXXWAMHA.js';
import './chunks/chunk-E5QSZH7H.js';
import './chunks/chunk-MRMVJNCY.js';
import './chunks/chunk-GJ4TBRAP.js';
import './chunks/chunk-4ZBW3RPJ.js';
import './chunks/chunk-NFNIXDBO.js';
import './chunks/chunk-3VBPBP2I.js';
import './chunks/chunk-TDA2CSOA.js';
import './chunks/chunk-TR4IJQJH.js';
import './chunks/chunk-DLGVRSHY.js';
import './chunks/chunk-TQCGHYMQ.js';
import './chunks/chunk-UQKO4MPH.js';
import './chunks/chunk-DULVJHTP.js';
import './chunks/chunk-VZDRI75F.js';
import './chunks/chunk-IMMXIVEN.js';
import './chunks/chunk-PFFUSUZV.js';
import './chunks/chunk-ZUA4UCCK.js';
import './chunks/chunk-E5TN64FJ.js';
import './chunks/chunk-C2JFZAA6.js';
import './chunks/chunk-BMAX7PVF.js';
import './chunks/chunk-BIGBL5GY.js';
import './chunks/chunk-WAHVE4HI.js';
import './chunks/chunk-UHV2CLK4.js';
import './chunks/chunk-LKBGYH24.js';
import './chunks/chunk-NPAVUIYS.js';
import './chunks/chunk-5WQWCIW2.js';
import './chunks/chunk-XVTO5FXB.js';
import './chunks/chunk-TXEI3VIO.js';
import './chunks/chunk-IF5A3X5R.js';
import './chunks/chunk-A56CZ36T.js';
import './chunks/chunk-VCZJ3L43.js';
import './chunks/chunk-KMNJXPZW.js';
import './chunks/chunk-IWQIFRAQ.js';
import './chunks/chunk-IXJUZ6BJ.js';
import './chunks/chunk-6RYECJ22.js';
import './chunks/chunk-H35ZN5RF.js';
import './chunks/chunk-SFQIIX7V.js';
import './chunks/chunk-4SREEBHL.js';
import './chunks/chunk-LH7BV2XD.js';
import './chunks/chunk-HUOGO4QD.js';
import './chunks/chunk-ICAJZZ55.js';
import './chunks/chunk-6YPVUWVV.js';
import './chunks/chunk-RLGAJ5QB.js';
import './chunks/chunk-VPSTUDLO.js';
import { isString, listenEvent, DOMEvent } from 'maverick.js/std';
export { appendTriggerEvent, findTriggerEvent, hasTriggerEvent, walkTriggerEventChain } from 'maverick.js/std';

var LibASSTextRenderer = class {
  constructor(loader, config) {
    this.loader = loader;
    this.config = config;
    this.priority = 1;
    this._instance = null;
    this._track = null;
    this._typeRE = /(ssa|ass)$/;
  }
  canRender(track) {
    return !!track.src && (isString(track.type) && this._typeRE.test(track.type) || this._typeRE.test(track.src));
  }
  attach(video) {
    this.loader().then(async (mod) => {
      var _a;
      this._instance = new mod.default({
        ...this.config,
        video,
        subUrl: ((_a = this._track) == null ? void 0 : _a.src) || ""
      });
      listenEvent(this._instance, "ready", () => {
        var _a2;
        const canvas = (_a2 = this._instance) == null ? void 0 : _a2._canvas;
        if (canvas)
          canvas.style.pointerEvents = "none";
      });
      listenEvent(this._instance, "error", (event) => {
        if (this._track) {
          this._track[TEXT_TRACK_READY_STATE] = 3;
          this._track.dispatchEvent(
            new DOMEvent("error", {
              trigger: event,
              detail: event.error
            })
          );
        }
      });
    });
  }
  changeTrack(track) {
    var _a;
    if (!track || track.readyState === 3) {
      this._freeTrack();
    } else if (this._track !== track) {
      (_a = this._instance) == null ? void 0 : _a.setTrackByUrl(track.src);
      this._track = track;
    }
  }
  detach() {
    this._freeTrack();
  }
  _freeTrack() {
    var _a;
    (_a = this._instance) == null ? void 0 : _a.freeTrack();
    this._track = null;
  }
};

export { LibASSTextRenderer };
