// vidstack@0.6.9/dist/prod/chunks/chunk-42OM7QJT.js
import { lazyPaths } from './chunk-R6FSMW74.js';
import { $$_create_walker, $$_effect, $$_hydrating, $$_create_template } from 'maverick.js/dom';
import { signal, effect, peek } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><svg viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" data-media-icon="true"></svg>`);
var Icon = class extends Component {
  constructor() {
    super(...arguments);
    this.ed = false;
    this.fd = signal("");
  }
  onAttach(el) {
    this.ed = el.hasAttribute("mk-h");
    effect(this.qi.bind(this));
  }
  qi() {
    const type = this.$props.type();
    if (this.ed) {
      this.ed = false;
      return;
    }
    if (type && lazyPaths[type]) {
      lazyPaths[type]().then(({ default: paths2 }) => {
        if (type === peek(this.$props.type))
          this.fd.set(paths2);
      });
    } else {
      this.fd.set("");
    }
  }
  render() {
    return (() => {
      const [$$_root, $$_walker] = $$_create_walker($$_templ);
      $$_effect(() => {
        if (!$$_hydrating)
          $$_root.innerHTML = this.fd();
      });
      return $$_root;
    })();
  }
};
Icon.el = defineElement({
  tagName: "media-icon",
  props: { type: void 0 }
});

export { Icon };
