// vidstack@0.6.9/dist/prod/chunks/chunk-4DGDZPR6.js
import { radioGroupContext } from './chunk-4L7R5OAB.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { onPress } from './chunk-HQQTSFXD.js';
import { menuContext } from './chunk-HSL7QSWM.js';
import { FocusVisibleController } from './chunk-UMWH4QNT.js';
import { __decorateClass } from './chunk-VPSTUDLO.js';
import { $$_next_template, $$_create_template } from 'maverick.js/dom';
import { signal, effect, hasProvidedContext, useContext, peek } from 'maverick.js';
import { defineElement, prop, Component } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><div part="check"></div>`);
var Radio = class extends Component {
  constructor(instance) {
    super(instance);
    this.va = signal(false);
    this.E = {
      _: this.$props.value,
      lb: this.lb.bind(this),
      Ib: null
    };
    new FocusVisibleController(instance);
  }
  get checked() {
    return this.va();
  }
  onAttach() {
    effect(this.yc.bind(this));
    this.ze();
    const isMenuItem = hasProvidedContext(menuContext);
    this.setAttributes({
      value: this.$props.value,
      role: isMenuItem ? "menuitemradio" : "radio",
      tabindex: isMenuItem ? -1 : 0,
      "aria-checked": $ariaBool(this.va)
    });
  }
  onConnect(el) {
    this.ze();
    onPress(el, this.J.bind(this));
  }
  onDisconnect() {
    const group = useContext(radioGroupContext);
    group.remove(this.E);
  }
  ze() {
    const group = useContext(radioGroupContext);
    group.add(this.E);
  }
  yc() {
    const { value } = this.$props, newValue = value();
    if (peek(this.va)) {
      this.E.Ib?.(newValue);
    }
  }
  J(event) {
    if (peek(this.va))
      return;
    this.va.set(true);
    this.dispatch("change", { trigger: event });
    this.E.Ib?.(peek(this.$props.value), event);
  }
  lb(value, trigger) {
    if (peek(this.va) === value)
      return;
    this.va.set(value);
    this.dispatch("change", { trigger });
  }
  render() {
    return $$_next_template($$_templ);
  }
};
Radio.el = defineElement({
  tagName: "media-radio",
  props: { value: "" }
});
__decorateClass([
  prop
], Radio.prototype, "checked", 1);

export { Radio };
