// vidstack@0.6.9/dist/prod/chunks/chunk-332RFI7E.js
import { setAttributeIfEmpty, onPress } from './chunk-HQQTSFXD.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { TooltipController } from './chunk-QOO3T2DC.js';
import { FocusVisibleController } from './chunk-UMWH4QNT.js';
import { __decorateClass } from './chunk-VPSTUDLO.js';
import { signal, peek, isWriteSignal, onDispose } from 'maverick.js';
import { defineElement, prop, Component, ComponentController } from 'maverick.js/element';
import { ariaBool } from 'maverick.js/std';

var ARIAKeyShortcuts = class extends ComponentController {
  constructor(instance, _shortcut) {
    super(instance);
    this.pc = _shortcut;
  }
  onAttach(el) {
    const { $props, ariaKeys } = useMedia(), keys = el.getAttribute("aria-keyshortcuts");
    if (keys) {
      ariaKeys[this.pc] = keys;
      {
        onDispose(() => {
          delete ariaKeys[this.pc];
        });
      }
      return;
    }
    const shortcuts = $props.keyShortcuts()[this.pc];
    if (shortcuts)
      el.setAttribute("aria-keyshortcuts", shortcuts);
  }
};

// src/player/ui/buttons/toggle-button.tsx
var toggleButtonProps = {
  disabled: false,
  defaultPressed: false,
  defaultAppearance: false
};
var ToggleButton = class extends Component {
  constructor(instance) {
    super(instance);
    this.v = signal(false);
    new FocusVisibleController(instance);
    new TooltipController(instance);
    if (this.pa)
      new ARIAKeyShortcuts(instance, this.pa);
  }
  get pressed() {
    return peek(this.v);
  }
  onAttach(el) {
    if (isWriteSignal(this.v)) {
      this.v.set(this.$props.defaultPressed());
    }
    setAttributeIfEmpty(el, "tabindex", "0");
    setAttributeIfEmpty(el, "role", "button");
    const { disabled, defaultAppearance } = this.$props;
    this.setAttributes({
      disabled,
      "default-appearance": defaultAppearance,
      "data-pressed": this.v,
      "aria-pressed": this.Cg.bind(this),
      "data-media-button": true
    });
  }
  onConnect(el) {
    onPress(el, this.Dg.bind(this));
  }
  Cg() {
    return ariaBool(this.v());
  }
  J(event) {
    if (isWriteSignal(this.v)) {
      this.v.set((p) => !p);
    }
  }
  Dg(event) {
    const disabled = this.$props.disabled();
    if (disabled) {
      if (disabled)
        event.stopImmediatePropagation();
      return;
    }
    event.preventDefault();
    this.J(event);
  }
};
ToggleButton.el = defineElement({
  tagName: "media-toggle-button",
  props: toggleButtonProps
});
__decorateClass([
  prop
], ToggleButton.prototype, "pressed", 1);

export { ToggleButton, toggleButtonProps };
