// vidstack@0.6.9/dist/prod/chunks/chunk-T4T7THLQ.js
import { renderRadioGroup } from './chunk-KRJ7Q224.js';
import { ClassManager } from './chunk-P3PWPLZT.js';
import { RadioGroup } from './chunk-P3YNMH2O.js';
import { Radio } from './chunk-4DGDZPR6.js';
import { MenuItems } from './chunk-QGM7SQHY.js';
import { isTrackCaptionKind } from './chunk-PXCLVUYU.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_insert_at_marker_lite, $$_create_template } from 'maverick.js/dom';
import { effect } from 'maverick.js';
import { defineElement } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><span part="label"><!$></span>`);
var $$_templ_2 = $$_templ;
var CaptionsMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this.a = useMedia();
  }
  onConnect(el) {
    effect(this.wa.bind(this));
    effect(this.xa.bind(this));
    const { radioClass, radioCheckClass } = this.$props;
    new ClassManager(el).u("media-radio", radioClass).u('[part="check"]', radioCheckClass);
  }
  xa() {
    const { offLabel } = this.$props, { textTrack } = this.a.$store, track = textTrack();
    this.A.kb.set(
      track && isTrackCaptionKind(track) && track.mode === "showing" ? track.label : offLabel()
    );
  }
  wa() {
    this.A.ua(this.p());
  }
  p() {
    const { textTracks } = this.a.$store;
    return textTracks().filter(isTrackCaptionKind).length === 0;
  }
  o(event) {
    if (this.p())
      return;
    const radioGroup = event.target;
    const value = radioGroup.value;
    if (value === "off") {
      const track = this.a.textTracks.selected;
      if (track) {
        const index2 = this.a.textTracks.toArray().indexOf(track);
        this.a.remote.changeTextTrackMode(index2, "disabled", event);
      }
      return;
    }
    const index = this.a.textTracks.toArray().findIndex((track) => track.label.toLowerCase() === value);
    if (index >= 0)
      this.a.remote.changeTextTrackMode(index, "showing", event);
  }
  $() {
    const { textTrack, textTracks } = this.a.$store, track = textTrack();
    return track && isTrackCaptionKind(track) && track.mode === "showing" ? track.label.toLowerCase() : "off";
  }
  aa() {
    const { offLabel } = this.$props, { textTracks } = this.a.$store;
    return [
      { value: "off", content: () => (() => {
        const [$$_root, $$_walker] = $$_create_walker($$_templ), $$_expr = $$_walker.nextNode();
        $$_insert_at_marker_lite($$_expr, offLabel);
        return $$_root;
      })() },
      ...textTracks().filter(isTrackCaptionKind).map((track) => ({
        value: track.label.toLowerCase(),
        content: () => (() => {
          const [$$_root, $$_walker] = $$_create_walker($$_templ_2), $$_expr = $$_walker.nextNode();
          $$_insert_at_marker_lite($$_expr, track.label);
          return $$_root;
        })()
      }))
    ];
  }
  render() {
    const { radioGroupClass } = this.$props;
    return renderRadioGroup({
      value: this.$.bind(this),
      options: this.aa.bind(this),
      radioGroupClass,
      onChange: this.o.bind(this)
    });
  }
};
CaptionsMenuItems.el = defineElement({
  tagName: "media-captions-menu-items",
  props: {
    offLabel: "Off",
    radioClass: null,
    radioGroupClass: null,
    radioCheckClass: null
  }
});
CaptionsMenuItems.register = [RadioGroup, Radio];

export { CaptionsMenuItems };
