// vidstack@0.6.9/dist/prod/chunks/chunk-YBFM664V.js
import { VideoProvider } from './chunk-F3MPEIU4.js';
import { RAFLoop, LIST_SELECT, LIST_ADD } from './chunk-QSX4JG34.js';
import { isHLSSupported, IS_CHROME } from './chunk-WSYKLWQE.js';
import { TextTrack, TEXT_TRACK_READY_STATE, TEXT_TRACK_ON_MODE_CHANGE } from './chunk-PXCLVUYU.js';
import { preconnect, loadScript } from './chunk-IVW6GMZV.js';
import { peek, effect } from 'maverick.js';
import { isString, DOMEvent, listenEvent, isUndefined, isFunction, camelToKebabCase } from 'maverick.js/std';

// src/player/core/quality/symbols.ts
var SET_AUTO_QUALITY = Symbol(0);
var ENABLE_AUTO_QUALITY = Symbol(0);

// src/player/core/providers/hls/hls.ts
var toDOMEventType = (type) => camelToKebabCase(type);
var HLSController = class {
  constructor(_video) {
    this.k = _video;
    this.i = null;
    this.Wb = null;
    this.Xb = {};
    this.Yb = /* @__PURE__ */ new Set();
  }
  get instance() {
    return this.i;
  }
  setup(ctor, context) {
    this.c = context;
    const isLive = peek(context.$store.streamType).includes("live"), isLiveLowLatency = peek(context.$store.streamType).includes("ll-");
    this.i = new ctor({
      lowLatencyMode: isLiveLowLatency,
      backBufferLength: isLiveLowLatency ? 4 : isLive ? 8 : void 0,
      renderTextTracksNatively: false,
      ...this.Xb
    });
    const dispatcher = this.qf.bind(this);
    for (const event of Object.values(ctor.Events))
      this.i.on(event, dispatcher);
    this.i.on(ctor.Events.ERROR, this.la.bind(this));
    for (const callback of this.Yb)
      callback(this.i);
    context.player.dispatchEvent(new DOMEvent("hls-instance", { detail: this.i }));
    this.i.attachMedia(this.k);
    this.i.on(ctor.Events.AUDIO_TRACK_SWITCHED, this.rf.bind(this));
    this.i.on(ctor.Events.LEVEL_SWITCHED, this.sf.bind(this));
    this.i.on(ctor.Events.LEVEL_LOADED, this.tf.bind(this));
    this.i.on(ctor.Events.NON_NATIVE_TEXT_TRACKS_FOUND, this.uf.bind(this));
    this.i.on(ctor.Events.CUES_PARSED, this.vf.bind(this));
    context.qualities[ENABLE_AUTO_QUALITY] = this.wf.bind(this);
    listenEvent(context.qualities, "change", this.Zb.bind(this));
    listenEvent(context.audioTracks, "change", this.xf.bind(this));
    this.Wb = effect(this.yf.bind(this));
  }
  yf() {
    if (!this.c.$store.live())
      return;
    const raf = new RAFLoop(this.zf.bind(this));
    raf.pb();
    return raf.ia.bind(raf);
  }
  zf() {
    this.c.$store.liveSyncPosition.set(this.i?.liveSyncPosition ?? Infinity);
  }
  qf(eventType, detail) {
    this.c.player.dispatchEvent(new DOMEvent(toDOMEventType(eventType), { detail }));
  }
  uf(eventType, data) {
    const event = new DOMEvent(eventType, { detail: data });
    let currentTrack = -1;
    for (let i = 0; i < data.tracks.length; i++) {
      const nonNativeTrack = data.tracks[i], init = nonNativeTrack.subtitleTrack ?? nonNativeTrack.closedCaptions, track = new TextTrack({
        id: `hls-${nonNativeTrack.kind}${i}`,
        src: init?.url,
        label: nonNativeTrack.label,
        language: init?.lang,
        kind: nonNativeTrack.kind
      });
      track[TEXT_TRACK_READY_STATE] = 2;
      track[TEXT_TRACK_ON_MODE_CHANGE] = () => {
        if (track.mode === "showing") {
          this.i.subtitleTrack = i;
          currentTrack = i;
        } else if (currentTrack === i) {
          this.i.subtitleTrack = -1;
          currentTrack = -1;
        }
      };
      if (nonNativeTrack.default)
        track.setMode("showing", event);
      this.c.textTracks.add(track, event);
    }
  }
  vf(eventType, data) {
    const track = this.c.textTracks.getById(`hls-${data.track}`);
    if (!track)
      return;
    const event = new DOMEvent(eventType, { detail: data });
    for (const cue of data.cues) {
      cue.positionAlign = "auto";
      track.addCue(cue, event);
    }
  }
  rf(eventType, data) {
    const track = this.c.audioTracks[data.id];
    if (track) {
      this.c.audioTracks[LIST_SELECT](
        track,
        true,
        new DOMEvent(eventType, { detail: data })
      );
    }
  }
  sf(eventType, data) {
    const quality = this.c.qualities[data.level];
    if (quality) {
      this.c.qualities[LIST_SELECT](
        quality,
        true,
        new DOMEvent(eventType, { detail: data })
      );
    }
  }
  tf(eventType, data) {
    if (this.c.$store.canPlay())
      return;
    const { type, live, totalduration: duration } = data.details;
    const event = new DOMEvent(eventType, { detail: data });
    this.c.delegate.d("stream-type-change", {
      detail: live ? type === "EVENT" && Number.isFinite(duration) ? "live:dvr" : "live" : "on-demand",
      trigger: event
    });
    this.c.delegate.d("duration-change", { detail: duration, trigger: event });
    const media = this.i.media;
    if (this.i.currentLevel === -1) {
      this.c.qualities[SET_AUTO_QUALITY](true, event);
    }
    for (const track of this.i.audioTracks) {
      this.c.audioTracks[LIST_ADD](
        {
          id: track.id + "",
          label: track.name,
          language: track.lang || "",
          kind: "main"
        },
        event
      );
    }
    for (const level of this.i.levels) {
      this.c.qualities[LIST_ADD](
        {
          width: level.width,
          height: level.height,
          codec: level.codecSet,
          bitrate: level.bitrate
        },
        event
      );
    }
    media.dispatchEvent(new DOMEvent("canplay", { trigger: event }));
  }
  la(eventType, data) {
    if (data.fatal) {
      switch (data.type) {
        case "networkError":
          this.i?.startLoad();
          break;
        case "mediaError":
          this.i?.recoverMediaError();
          break;
        default:
          this.i?.destroy();
          this.i = null;
          break;
      }
    }
  }
  wf() {
    if (this.i)
      this.i.currentLevel = -1;
  }
  Zb() {
    const { qualities } = this.c;
    if (!this.i || qualities.auto)
      return;
    this.i[qualities.switch + "Level"] = qualities.selectedIndex;
    if (IS_CHROME)
      this.k.currentTime = this.k.currentTime;
  }
  xf() {
    const { audioTracks } = this.c;
    if (this.i && this.i.audioTrack !== audioTracks.selectedIndex) {
      this.i.audioTrack = audioTracks.selectedIndex;
    }
  }
  _b() {
    if (this.c)
      this.c.qualities[ENABLE_AUTO_QUALITY] = void 0;
    this.i?.destroy();
    this.i = null;
    this.Wb?.();
    this.Wb = null;
  }
};

// src/utils/error.ts
function coerceToError(error) {
  return error instanceof Error ? error : Error(JSON.stringify(error));
}

// src/player/core/providers/hls/lib-loader.ts
var HLSLibLoader = class {
  constructor(_lib, _context, _callback) {
    this.ac = _lib;
    this.c = _context;
    this.ja = _callback;
    this.Af();
  }
  async Af() {
    const callbacks = {
      onLoadStart: this.ka.bind(this),
      onLoaded: this.$b.bind(this),
      onLoadError: this.Bf.bind(this)
    };
    let ctor = await loadHLSScript(this.ac, callbacks);
    if (isUndefined(ctor) && !isString(this.ac))
      ctor = await importHLS(this.ac, callbacks);
    if (!ctor)
      return null;
    if (!ctor.isSupported()) {
      const message = "[vidstack]: `hls.js` is not supported in this environment";
      this.c.player.dispatchEvent(new DOMEvent("hls-unsupported"));
      this.c.delegate.d("error", { detail: { message, code: 4 } });
      return null;
    }
    return ctor;
  }
  ka() {
    this.c.player.dispatchEvent(new DOMEvent("hls-lib-load-start"));
  }
  $b(ctor) {
    this.c.player.dispatchEvent(
      new DOMEvent("hls-lib-loaded", {
        detail: ctor
      })
    );
    this.ja(ctor);
  }
  Bf(e) {
    const error = coerceToError(e);
    this.c.player.dispatchEvent(
      new DOMEvent("hls-lib-load-error", {
        detail: error
      })
    );
    this.c.delegate.d("error", {
      detail: { message: error.message, code: 4 }
    });
  }
};
async function importHLS(loader, callbacks = {}) {
  if (isUndefined(loader))
    return void 0;
  callbacks.onLoadStart?.();
  if (loader.prototype && loader.prototype !== Function) {
    callbacks.onLoaded?.(loader);
    return loader;
  }
  try {
    const ctor = (await loader())?.default;
    if (ctor && !!ctor.isSupported) {
      callbacks.onLoaded?.(ctor);
    } else {
      throw Error(
        false ? "[vidstack] failed importing `hls.js`. Dynamic import returned invalid constructor." : ""
      );
    }
    return ctor;
  } catch (err) {
    callbacks.onLoadError?.(err);
  }
  return void 0;
}
async function loadHLSScript(src, callbacks = {}) {
  if (!isString(src))
    return void 0;
  callbacks.onLoadStart?.();
  try {
    await loadScript(src);
    if (!isFunction(window.Hls)) {
      throw Error(
        false ? "[vidstack] failed loading `hls.js`. Could not find a valid `Hls` constructor on window" : ""
      );
    }
    const ctor = window.Hls;
    callbacks.onLoaded?.(ctor);
    return ctor;
  } catch (err) {
    callbacks.onLoadError?.(err);
  }
  return void 0;
}

// src/player/core/providers/hls/provider.ts
var HLS_PROVIDER = Symbol(0);
var JS_DELIVR_CDN = "https://cdn.jsdelivr.net";
var _a;
var HLSProvider = class extends VideoProvider {
  constructor() {
    super(...arguments);
    this[_a] = true;
    this.yd = null;
    this.E = new HLSController(this.video);
    this.Za = `${JS_DELIVR_CDN}/npm/hls.js@^1.0.0/dist/hls${".min.js"}`;
  }
  /**
   * The `hls.js` constructor.
   */
  get ctor() {
    return this.yd;
  }
  /**
   * The current `hls.js` instance.
   */
  get instance() {
    return this.E.instance;
  }
  get type() {
    return "hls";
  }
  get canLiveSync() {
    return true;
  }
  /**
   * The `hls.js` configuration object.
   *
   * @see {@link https://github.com/video-dev/hls.js/blob/master/docs/API.md#fine-tuning}
   */
  get config() {
    return this.E.Xb;
  }
  set config(config) {
    this.E.Xb = config;
  }
  /**
   * The `hls.js` constructor (supports dynamic imports) or a URL of where it can be found.
   *
   * @defaultValue `https://cdn.jsdelivr.net/npm/hls.js@^1.0.0/dist/hls.min.js`
   */
  get library() {
    return this.Za;
  }
  set library(library) {
    this.Za = library;
  }
  preconnect() {
    if (!isString(this.Za))
      return;
    preconnect(this.Za);
  }
  setup(context) {
    super.setup(context);
    new HLSLibLoader(this.Za, context, (ctor) => {
      this.yd = ctor;
      this.E.setup(ctor, context);
      context.delegate.d("provider-setup", { detail: this });
      const src = peek(context.$store.source);
      if (src)
        this.loadSource(src);
    });
  }
  async loadSource({ src }) {
    if (!isString(src))
      return;
    this.E.instance?.loadSource(src);
  }
  /**
   * The given callback is invoked when a new `hls.js` instance is created and right before it's
   * attached to media.
   */
  onInstance(callback) {
    const instance = this.E.instance;
    if (instance)
      callback(instance);
    this.E.Yb.add(callback);
    return () => this.E.Yb.delete(callback);
  }
  destroy() {
    this.E._b();
  }
};
_a = HLS_PROVIDER;
/**
 * Whether `hls.js` is supported in this environment.
 */
HLSProvider.supported = isHLSSupported();

export { ENABLE_AUTO_QUALITY, HLSProvider, HLS_PROVIDER, SET_AUTO_QUALITY, coerceToError };
