// vidstack@0.6.9/dist/prod/chunks/chunk-EM6EO3FW.js
import { preconnect } from './chunk-IVW6GMZV.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_create_walker, $$_effect, $$_attr, $$_listen, $$_create_template } from 'maverick.js/dom';
import { signal, computed, effect } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';

var $$_templ = /* @__PURE__ */ $$_create_template(`<!$><img part="img" />`);
var Poster = class extends Component {
  constructor() {
    super(...arguments);
    this.Jb = signal(true);
    this.cd = signal(false);
  }
  onAttach(el) {
    this.a = useMedia();
    this.dd = computed(this.ji.bind(this));
    this.Ee = this.ki.bind(this);
    this.setAttributes({
      "data-loading": this.Jb,
      "aria-hidden": $ariaBool(this.z.bind(this))
    });
  }
  onConnect(el) {
    const { canLoad, poster } = this.a.$store;
    window.requestAnimationFrame(() => {
      if (!canLoad())
        preconnect(poster());
    });
    effect(this.ka.bind(this));
  }
  z() {
    const { poster } = this.a.$store;
    return this.cd() || !poster();
  }
  ji() {
    const { canLoad, poster } = this.a.$store;
    return canLoad() && poster().length ? poster() : null;
  }
  ki() {
    return this.dd() ? this.$props.alt() : null;
  }
  ka() {
    const { canLoad, poster } = this.a.$store;
    const isLoading = canLoad() && !!poster();
    this.Jb.set(isLoading);
    this.cd.set(false);
  }
  li() {
    this.Jb.set(false);
  }
  la() {
    this.Jb.set(false);
    this.cd.set(true);
  }
  render() {
    const { crossorigin } = this.a.$store;
    return (() => {
      const [$$_root, $$_walker] = $$_create_walker($$_templ);
      $$_effect(() => $$_attr($$_root, "src", this.dd()));
      $$_effect(() => $$_attr($$_root, "alt", this.Ee()));
      $$_effect(() => $$_attr($$_root, "crossorigin", crossorigin()));
      $$_listen($$_root, "load", this.li.bind(this));
      $$_listen($$_root, "error", this.la.bind(this));
      return $$_root;
    })();
  }
};
Poster.el = defineElement({
  tagName: "media-poster",
  props: { alt: void 0 }
});

export { Poster };
