// vidstack@0.6.9/dist/prod/chunks/chunk-F3MPEIU4.js
import { HTMLMediaProvider } from './chunk-QSX4JG34.js';
import { canUseVideoPresentation, canUsePictureInPicture, canPlayHLSNatively } from './chunk-WSYKLWQE.js';
import { ATTACH_VIDEO, TextTrack, TEXT_TRACK_NATIVE, TEXT_TRACK_READY_STATE, TEXT_TRACK_NATIVE_HLS } from './chunk-PXCLVUYU.js';
import { onDispose } from 'maverick.js';
import { listenEvent, DOMEvent } from 'maverick.js/std';

var NativeHLSTextTracks = class {
  constructor(_video, _context) {
    this.k = _video;
    this.c = _context;
    _video.textTracks.onaddtrack = this.Vb.bind(this);
    onDispose(this.Rb.bind(this));
  }
  Vb(event2) {
    const nativeTrack = event2.track;
    if (!nativeTrack || findTextTrackElement(this.k, nativeTrack))
      return;
    const track = new TextTrack({
      id: nativeTrack.id,
      kind: nativeTrack.kind,
      label: nativeTrack.label,
      language: nativeTrack.language,
      type: "vtt"
    });
    track[TEXT_TRACK_NATIVE] = { track: nativeTrack };
    track[TEXT_TRACK_READY_STATE] = 2;
    track[TEXT_TRACK_NATIVE_HLS] = true;
    let lastIndex = 0;
    const onCueChange = (event3) => {
      if (!nativeTrack.cues)
        return;
      for (let i = lastIndex; i < nativeTrack.cues.length; i++) {
        track.addCue(nativeTrack.cues[i], event3);
        lastIndex++;
      }
    };
    onCueChange(event2);
    nativeTrack.oncuechange = onCueChange;
    this.c.textTracks.add(track, event2);
    track.setMode(nativeTrack.mode, event2);
  }
  Rb() {
    this.k.textTracks.onaddtrack = null;
    for (const track of this.c.textTracks) {
      const nativeTrack = track[TEXT_TRACK_NATIVE]?.track;
      if (nativeTrack?.oncuechange)
        nativeTrack.oncuechange = null;
    }
  }
};
function findTextTrackElement(video, track) {
  return Array.from(video.children).find((el) => el.track === track);
}
var VideoPictureInPicture = class {
  constructor(_video, _media) {
    this.k = _video;
    this.a = _media;
    this.o = (active, event2) => {
      this.a.delegate.d("picture-in-picture-change", {
        detail: active,
        trigger: event2
      });
    };
    listenEvent(this.k, "enterpictureinpicture", this.nf.bind(this));
    listenEvent(this.k, "leavepictureinpicture", this.of.bind(this));
  }
  get active() {
    return document.pictureInPictureElement === this.k;
  }
  get supported() {
    return canUsePictureInPicture(this.k);
  }
  async enter() {
    return this.k.requestPictureInPicture();
  }
  exit() {
    return document.exitPictureInPicture();
  }
  nf(event2) {
    this.o(true, event2);
  }
  of(event2) {
    this.o(false, event2);
  }
};
var VideoPresentation = class {
  constructor(_video, _media) {
    this.k = _video;
    this.a = _media;
    this.F = "inline";
    listenEvent(this.k, "webkitpresentationmodechanged", this.pf.bind(this));
  }
  get xd() {
    return canUseVideoPresentation(this.k);
  }
  async tb(mode) {
    if (this.F === mode)
      return;
    this.k.webkitSetPresentationMode(mode);
  }
  pf() {
    const prevMode = this.F;
    this.F = this.k.webkitPresentationMode;
    this.a.player?.dispatchEvent(
      new DOMEvent("video-presentation-change", {
        detail: this.F,
        trigger: event
      })
    );
    ["fullscreen", "picture-in-picture"].forEach((type) => {
      if (this.F === type || prevMode === type) {
        this.a.delegate.d(`${type}-change`, {
          detail: this.F === type,
          trigger: event
        });
      }
    });
  }
};
var FullscreenPresentationAdapter = class {
  constructor(_presentation) {
    this.U = _presentation;
  }
  get active() {
    return this.U.F === "fullscreen";
  }
  get supported() {
    return this.U.xd;
  }
  async enter() {
    this.U.tb("fullscreen");
  }
  async exit() {
    this.U.tb("inline");
  }
};
var PIPPresentationAdapter = class {
  constructor(_presentation) {
    this.U = _presentation;
  }
  get active() {
    return this.U.F === "picture-in-picture";
  }
  get supported() {
    return this.U.xd;
  }
  async enter() {
    this.U.tb("picture-in-picture");
  }
  async exit() {
    this.U.tb("inline");
  }
};

// src/player/core/providers/video/provider.ts
var VIDEO_PROVIDER = Symbol(0);
var _a;
var VideoProvider = class extends HTMLMediaProvider {
  constructor(video, context) {
    super(video);
    this[_a] = true;
    if (canUseVideoPresentation(video)) {
      const presentation = new VideoPresentation(video, context);
      this.fullscreen = new FullscreenPresentationAdapter(presentation);
      this.pictureInPicture = new PIPPresentationAdapter(presentation);
    } else if (canUsePictureInPicture(video)) {
      this.pictureInPicture = new VideoPictureInPicture(video, context);
    }
  }
  get type() {
    return "video";
  }
  setup(context) {
    super.setup(context);
    if (canPlayHLSNatively(this.video)) {
      new NativeHLSTextTracks(this.video, context);
    }
    context.textRenderers[ATTACH_VIDEO](this.video);
    onDispose(() => {
      context.textRenderers[ATTACH_VIDEO](null);
    });
    if (this.type === "video")
      context.delegate.d("provider-setup", { detail: this });
  }
  /**
   * The native HTML `<video>` element.
   *
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement}
   */
  get video() {
    return this.a;
  }
};
_a = VIDEO_PROVIDER;

export { VIDEO_PROVIDER, VideoProvider };
