// vidstack@0.6.9/dist/prod/chunks/chunk-PXCLVUYU.js
import { getRequestCredentials } from './chunk-IVW6GMZV.js';
import { isCueActive } from './chunk-NOPW2CG3.js';
import { EventsTarget, DOMEvent, isNumber } from 'maverick.js/std';

// src/player/core/tracks/text/symbols.ts
var ATTACH_VIDEO = Symbol(0);
var TEXT_TRACK_CROSSORIGIN = Symbol(0);
var TEXT_TRACK_READY_STATE = Symbol(0);
var TEXT_TRACK_UPDATE_ACTIVE_CUES = Symbol(0);
var TEXT_TRACK_CAN_LOAD = Symbol(0);
var TEXT_TRACK_ON_MODE_CHANGE = Symbol(0);
var TEXT_TRACK_NATIVE = Symbol(0);
var TEXT_TRACK_NATIVE_HLS = Symbol(0);
var _a, _b, _c;
var TextTrack = class extends EventsTarget {
  constructor(init) {
    super();
    this.id = "";
    this.label = "";
    this.language = "";
    this.default = false;
    this.Da = false;
    this.Ya = 0;
    this.F = "disabled";
    this.rd = {};
    this.sb = [];
    this.D = [];
    this.Ea = [];
    /* @internal */
    this[_a] = 0;
    /* @internal */
    this[_b] = null;
    /* @internal */
    this[_c] = null;
    for (const prop of Object.keys(init))
      this[prop] = init[prop];
    if (init.content) {
      import('media-captions').then(({ parseText, VTTCue, VTTRegion }) => {
        if (init.type === "json") {
          this.sd(init.content, VTTCue, VTTRegion);
        } else {
          parseText(init.content, { type: init.type }).then(({ cues, regions }) => {
            this.D = cues;
            this.sb = regions;
            this.td();
          });
        }
      });
    } else if (!init.src)
      this[TEXT_TRACK_READY_STATE] = 2;
  }
  static createId(track) {
    return `id::${track.type}-${track.kind}-${track.src ?? track.label}`;
  }
  get metadata() {
    return this.rd;
  }
  get regions() {
    return this.sb;
  }
  get cues() {
    return this.D;
  }
  get activeCues() {
    return this.Ea;
  }
  /**
   * - 0: Not Loading
   * - 1: Loading
   * - 2: Ready
   * - 3: Error
   */
  get readyState() {
    return this[TEXT_TRACK_READY_STATE];
  }
  get mode() {
    return this.F;
  }
  set mode(mode) {
    this.setMode(mode);
  }
  addCue(cue, trigger) {
    let i = 0, length = this.D.length;
    for (i = 0; i < length; i++)
      if (cue.endTime <= this.D[i].startTime)
        break;
    if (i === length)
      this.D.push(cue);
    else
      this.D.splice(i, 0, cue);
    if (trigger?.type !== "cuechange") {
      this[TEXT_TRACK_NATIVE]?.track.addCue(cue);
    }
    this.dispatchEvent(new DOMEvent("add-cue", { detail: cue, trigger }));
    if (isCueActive(cue, this.Ya)) {
      this[TEXT_TRACK_UPDATE_ACTIVE_CUES](this.Ya, trigger);
    }
  }
  removeCue(cue, trigger) {
    const index = this.D.indexOf(cue);
    if (index >= 0) {
      const isActive = this.Ea.includes(cue);
      this.D.splice(index, 1);
      this[TEXT_TRACK_NATIVE]?.track.removeCue(cue);
      this.dispatchEvent(new DOMEvent("remove-cue", { detail: cue, trigger }));
      if (isActive) {
        this[TEXT_TRACK_UPDATE_ACTIVE_CUES](this.Ya, trigger);
      }
    }
  }
  setMode(mode, trigger) {
    if (this.F === mode)
      return;
    this.F = mode;
    if (mode === "disabled") {
      this.Ea = [];
      this.ud();
    } else {
      this.vd();
    }
    this.dispatchEvent(new DOMEvent("mode-change", { detail: this, trigger }));
    this[TEXT_TRACK_ON_MODE_CHANGE]?.();
  }
  /* @internal */
  [(_a = TEXT_TRACK_READY_STATE, _b = TEXT_TRACK_ON_MODE_CHANGE, _c = TEXT_TRACK_NATIVE, TEXT_TRACK_UPDATE_ACTIVE_CUES)](currentTime, trigger) {
    this.Ya = currentTime;
    if (this.mode === "disabled" || !this.D.length)
      return;
    const activeCues = [];
    for (let i = 0, length = this.D.length; i < length; i++) {
      const cue = this.D[i];
      if (isCueActive(cue, currentTime))
        activeCues.push(cue);
    }
    let changed = activeCues.length !== this.Ea.length;
    if (!changed) {
      for (let i = 0; i < activeCues.length; i++) {
        if (!this.Ea.includes(activeCues[i])) {
          changed = true;
          break;
        }
      }
    }
    this.Ea = activeCues;
    if (changed)
      this.ud(trigger);
  }
  /* @internal */
  [TEXT_TRACK_CAN_LOAD]() {
    this.Da = true;
    if (this.F !== "disabled")
      this.vd();
  }
  async vd() {
    if (!this.Da || !this.src || this[TEXT_TRACK_READY_STATE] > 0)
      return;
    this[TEXT_TRACK_READY_STATE] = 1;
    this.dispatchEvent(new DOMEvent("load-start"));
    try {
      const { parseResponse, VTTCue, VTTRegion } = await import('media-captions'), crossorigin = this[TEXT_TRACK_CROSSORIGIN]?.();
      const response = fetch(this.src, {
        headers: this.type === "json" ? { "Content-Type": "application/json" } : void 0,
        credentials: getRequestCredentials(crossorigin)
      });
      if (this.type === "json") {
        this.sd(await (await response).text(), VTTCue, VTTRegion);
      } else {
        const { errors, metadata, regions, cues } = await parseResponse(response, {
          type: this.type,
          encoding: this.encoding
        });
        if (errors[0]?.code === 0) {
          throw errors[0];
        } else {
          this.rd = metadata;
          this.sb = regions;
          this.D = cues;
        }
      }
      this.td();
    } catch (error) {
      this.wd(error);
    }
  }
  td() {
    this[TEXT_TRACK_READY_STATE] = 2;
    const nativeTrack = this[TEXT_TRACK_NATIVE]?.track;
    if (nativeTrack)
      for (const cue of this.D)
        nativeTrack.addCue(cue);
    const loadEvent = new DOMEvent("load");
    this[TEXT_TRACK_UPDATE_ACTIVE_CUES](this.Ya, loadEvent);
    this.dispatchEvent(loadEvent);
  }
  wd(error) {
    this[TEXT_TRACK_READY_STATE] = 3;
    this.dispatchEvent(new DOMEvent("error", { detail: error }));
  }
  sd(json, VTTCue, VTTRegion) {
    try {
      json = JSON.parse(json);
      if (json.regions) {
        this.sb = json.regions.map((json2) => Object.assign(new VTTRegion(), json2));
      }
      if (json.cues) {
        this.D = json.cues.filter((json2) => isNumber(json2.startTime) && isNumber(json2.endTime)).map((json2) => Object.assign(new VTTCue(0, 0, ""), json2));
      }
    } catch (error) {
      this.wd(error);
    }
  }
  ud(trigger) {
    this.dispatchEvent(new DOMEvent("cue-change", { trigger }));
  }
};
var captionRE = /captions|subtitles/;
function isTrackCaptionKind(track) {
  return captionRE.test(track.kind);
}

export { ATTACH_VIDEO, TEXT_TRACK_CAN_LOAD, TEXT_TRACK_CROSSORIGIN, TEXT_TRACK_NATIVE, TEXT_TRACK_NATIVE_HLS, TEXT_TRACK_ON_MODE_CHANGE, TEXT_TRACK_READY_STATE, TEXT_TRACK_UPDATE_ACTIVE_CUES, TextTrack, isTrackCaptionKind };
