// vidstack/dist/prod/chunks/chunk-UHECQU42.js
import { signal, effect, onDispose, peek } from 'maverick.js';
import { animationFrameThrottle, isDOMElement, setAttribute } from 'maverick.js/std';

// src/foundation/observers/class-manager.ts
var ClassManager = class {
  constructor(el) {
    this.Hb = /* @__PURE__ */ new Map();
    this.Lc = signal([]);
    this.Mc = animationFrameThrottle(this.S.bind(this));
    this.Jc = el.firstChild;
    this.Kc = new MutationObserver(this.Cb.bind(this));
    this.Kc.observe(this.Jc, { subtree: true, childList: true });
    effect(this.jh.bind(this));
    onDispose(this.$b.bind(this));
  }
  Cb(records) {
    const selector = Array.from(this.Hb.keys()).join(",");
    for (const record of records) {
      for (const node of record.addedNodes) {
        if (isDOMElement(node) && node.matches(selector))
          this.Mc();
      }
    }
  }
  u(selector, $class) {
    this.Hb.set(selector, $class);
    this.Lc.set((c) => [...c, $class]);
    return this;
  }
  S() {
    for (const [selector, classes] of this.Hb) {
      const _class = peek(classes);
      for (const el of this.Jc.querySelectorAll(selector)) {
        setAttribute(el, "class", _class);
      }
    }
  }
  jh() {
    for (const c of this.Lc())
      c();
    this.Mc();
  }
  $b() {
    this.Lc.set([]);
    this.Hb.clear();
    this.Kc.disconnect();
  }
};

export { ClassManager };
