// vidstack/dist/prod/chunks/chunk-FIKNN6YM.js
import { sliderProps, Slider } from './chunk-LOOCLFQO.js';
import { SliderStoreFactory } from './chunk-53B74NMI.js';
import { functionThrottle } from './chunk-JELAJF2G.js';
import { canChangeVolume } from './chunk-WSYKLWQE.js';
import { round } from './chunk-CVLY5S52.js';
import { setAttributeIfEmpty } from './chunk-HQQTSFXD.js';
import { effect } from 'maverick.js';
import { defineElement, defineProp } from 'maverick.js/element';
import { setAttribute } from 'maverick.js/std';

var VolumeSlider = class extends Slider {
  constructor() {
    super(...arguments);
    this.xc = true;
    this.ke = functionThrottle(this.Tb.bind(this), 25);
  }
  onAttach(el) {
    setAttributeIfEmpty(el, "aria-label", "Media volume");
    super.onAttach(el);
    {
      canChangeVolume().then((canSet) => {
        if (!canSet)
          setAttribute(el, "aria-hidden", "true");
      });
    }
    effect(this.oc.bind(this));
  }
  oc() {
    const { muted, volume } = this.a.$store;
    const newValue = muted() ? 0 : volume() * 100;
    this.$store.value.set(newValue);
    this.dispatch("value-change", { detail: newValue });
  }
  Tb(event) {
    if (!event.trigger)
      return;
    const mediaVolume = round(event.detail / 100, 3);
    this.a.remote.changeVolume(mediaVolume, event);
  }
  tc(event) {
    this.ke(event);
  }
  uc(event) {
    this.ke(event);
  }
  Ac() {
    return 0;
  }
  Bc() {
    return 100;
  }
};
VolumeSlider.el = defineElement({
  tagName: "media-volume-slider",
  props: {
    ...sliderProps,
    min: defineProp({ value: 0, attribute: false }),
    max: defineProp({ value: 100, attribute: false }),
    value: defineProp({ value: 100, attribute: false })
  },
  store: SliderStoreFactory
});

export { VolumeSlider };
