// vidstack/dist/prod/chunks/chunk-DLSRK66Y.js
import { radioGroupContext } from './chunk-4L7R5OAB.js';
import { setAttributeIfEmpty } from './chunk-HQQTSFXD.js';
import { menuContext } from './chunk-HSL7QSWM.js';
import { __decorateClass } from './chunk-VPSTUDLO.js';
import { signal, provideContext, hasProvidedContext, effect, peek } from 'maverick.js';
import { defineElement, prop, Component } from 'maverick.js/element';

var RadioGroup = class extends Component {
  constructor(instance) {
    super(instance);
    this.Sa = /* @__PURE__ */ new Set();
    this._ = signal("");
    this.Zh = this.o.bind(this);
    provideContext(radioGroupContext, {
      add: this.Xh.bind(this),
      remove: this.Yh.bind(this)
    });
  }
  get values() {
    return Array.from(this.Sa).map((radio) => radio._());
  }
  get value() {
    return this._();
  }
  set value(value) {
    this.o(value);
  }
  onAttach(el) {
    const isMenuItem = hasProvidedContext(menuContext);
    if (!isMenuItem)
      setAttributeIfEmpty(el, "role", "radiogroup");
    this.ze();
    this.setAttributes({ value: this._ });
  }
  onConnect() {
    effect(this.ze.bind(this));
  }
  onDestroy() {
    this.Sa.clear();
  }
  Xh(radio) {
    if (this.Sa.has(radio))
      return;
    this.Sa.add(radio);
    radio.Jb = this.Zh;
    radio.mb(radio._() === this._());
  }
  Yh(radio) {
    radio.Jb = null;
    this.Sa.delete(radio);
  }
  ze() {
    this.o(this.$props.value());
  }
  o(newValue, trigger) {
    const currentValue = peek(this._);
    if (!newValue || newValue === currentValue)
      return;
    const currentRadio = this.Ae(currentValue), newRadio = this.Ae(newValue);
    currentRadio?.mb(false, trigger);
    newRadio?.mb(true, trigger);
    this._.set(newValue);
    this.dispatch("change", { trigger });
  }
  Ae(newValue) {
    for (const radio of this.Sa) {
      if (newValue === peek(radio._))
        return radio;
    }
    return null;
  }
};
RadioGroup.el = defineElement({
  tagName: "media-radio-group",
  props: { value: "" }
});
__decorateClass([
  prop
], RadioGroup.prototype, "values", 1);
__decorateClass([
  prop
], RadioGroup.prototype, "value", 1);

export { RadioGroup };
