// vidstack@0.6.9/dist/prod/index.js
export { CommunitySkin } from './chunks/chunk-ASNE53HT.js';
export { Tooltip } from './chunks/chunk-2IFPT6IV.js';
export { VolumeSlider } from './chunks/chunk-PPX44EJV.js';
export { SliderThumbnail } from './chunks/chunk-SVULT6MW.js';
export { SliderValue } from './chunks/chunk-ACF7J7V3.js';
export { SliderVideo } from './chunks/chunk-M2BNO6CW.js';
export { TimeSlider } from './chunks/chunk-LUMRDT57.js';
export { Slider, sliderProps } from './chunks/chunk-K5KNSMZO.js';
import './chunks/chunk-U43VNZZW.js';
export { SliderStoreFactory } from './chunks/chunk-53B74NMI.js';
export { Time } from './chunks/chunk-4Z3MBC6L.js';
export { PlaybackRateMenuItems } from './chunks/chunk-DPRRMVVS.js';
export { AudioTrackList, FullscreenController, List, MEDIA_KEY_SHORTCUTS, MediaRemoteControl, MediaStoreFactory, MediaUserController, Player, ScreenOrientationController, TextRenderers, TextTrackList, TimeRange, VideoQualityList, canFullscreen, getTimeRangesEnd, getTimeRangesStart, isAudioProvider, isHLSProvider, isHTMLAudioElement, isHTMLMediaElement, isHTMLVideoElement, isVideoProvider, softResetMediaStore } from './chunks/chunk-25SPJZP6.js';
import './chunks/chunk-IMVDTPPP.js';
import './chunks/chunk-YBFM664V.js';
import './chunks/chunk-F3MPEIU4.js';
import './chunks/chunk-QSX4JG34.js';
import './chunks/chunk-JELAJF2G.js';
export { Poster } from './chunks/chunk-EM6EO3FW.js';
export { QualityMenuButton } from './chunks/chunk-2E7XPR4Q.js';
export { QualityMenuItems } from './chunks/chunk-SIOMXY4Q.js';
export { SeekButton } from './chunks/chunk-RJENCUZN.js';
export { Menu } from './chunks/chunk-NGH77GR4.js';
export { MuteButton } from './chunks/chunk-QYNTGP7I.js';
export { Outlet } from './chunks/chunk-LXY5FWK3.js';
import './chunks/chunk-FRNAFKFA.js';
import './chunks/chunk-WSYKLWQE.js';
export { PIPButton } from './chunks/chunk-24P6MNPA.js';
export { PlayButton } from './chunks/chunk-PDW2Z3N4.js';
export { PlaybackRateMenuButton } from './chunks/chunk-IUG63NMT.js';
export { CaptionsMenuItems } from './chunks/chunk-T4T7THLQ.js';
export { Captions } from './chunks/chunk-WDQCF4QO.js';
export { ChaptersMenuItems } from './chunks/chunk-UVXZDTA7.js';
export { Thumbnail } from './chunks/chunk-ACCFVZCC.js';
import './chunks/chunk-JY53PGKC.js';
import './chunks/chunk-CVLY5S52.js';
export { FullscreenButton } from './chunks/chunk-FPZKPAJT.js';
export { Gesture } from './chunks/chunk-AU7E3EYY.js';
export { LiveIndicator } from './chunks/chunk-FDG4LG4U.js';
export { AudioMenuButton } from './chunks/chunk-PFYJAQB4.js';
export { AudioMenuItems } from './chunks/chunk-SBOQRVLH.js';
import './chunks/chunk-KRJ7Q224.js';
import './chunks/chunk-P3PWPLZT.js';
export { RadioGroup } from './chunks/chunk-P3YNMH2O.js';
export { Radio } from './chunks/chunk-4DGDZPR6.js';
import './chunks/chunk-4L7R5OAB.js';
export { MenuItems } from './chunks/chunk-QGM7SQHY.js';
export { BufferingIndicator } from './chunks/chunk-2KTXMEFW.js';
export { CaptionButton } from './chunks/chunk-MGAU6S2S.js';
import { TEXT_TRACK_READY_STATE } from './chunks/chunk-PXCLVUYU.js';
export { TextTrack, isTrackCaptionKind } from './chunks/chunk-PXCLVUYU.js';
import './chunks/chunk-IVW6GMZV.js';
import './chunks/chunk-NOPW2CG3.js';
export { ToggleButton, toggleButtonProps } from './chunks/chunk-332RFI7E.js';
import './chunks/chunk-EVP5E6AO.js';
import './chunks/chunk-HQQTSFXD.js';
export { mediaContext } from './chunks/chunk-ZIBTY3GT.js';
export { CaptionsMenuButton } from './chunks/chunk-MPX6QKCA.js';
import './chunks/chunk-HAOWXMSO.js';
export { MenuButton } from './chunks/chunk-CKDAUNQY.js';
import './chunks/chunk-HSL7QSWM.js';
import './chunks/chunk-OSS74BPE.js';
import './chunks/chunk-QOO3T2DC.js';
import './chunks/chunk-RBTESKWW.js';
import './chunks/chunk-R6FSMW74.js';
import './chunks/chunk-MF43X3Y3.js';
import './chunks/chunk-LBH32PRG.js';
import './chunks/chunk-KQUCP2EG.js';
import './chunks/chunk-DZHDFPQR.js';
import './chunks/chunk-Q2XL2SHG.js';
import './chunks/chunk-Z3R7IJO3.js';
import './chunks/chunk-2IK7NGIX.js';
import './chunks/chunk-KITPHWQZ.js';
import './chunks/chunk-HKZBS6I2.js';
import './chunks/chunk-WPVULLF5.js';
import './chunks/chunk-WNZBENRE.js';
import './chunks/chunk-TYOX6JVW.js';
import './chunks/chunk-ZUP6VWH2.js';
import './chunks/chunk-TT5YQ2JU.js';
import './chunks/chunk-3V72BDJY.js';
import './chunks/chunk-764WJ7YC.js';
import './chunks/chunk-J24QS7SA.js';
import './chunks/chunk-5AX42AEN.js';
import './chunks/chunk-I5ZNHUW2.js';
import './chunks/chunk-W4QNKYWE.js';
import './chunks/chunk-YGXIG4UH.js';
import './chunks/chunk-DJRMQYAX.js';
import './chunks/chunk-KU5TXDFG.js';
import './chunks/chunk-PDX6CZZG.js';
import './chunks/chunk-B37QG5XE.js';
import './chunks/chunk-RKG5SB2X.js';
import './chunks/chunk-IDM7ODL5.js';
import './chunks/chunk-CPWHUQBL.js';
import './chunks/chunk-EOTLQ6GG.js';
import './chunks/chunk-BQ2XLH4F.js';
import './chunks/chunk-E23QUJXR.js';
import './chunks/chunk-BM7OTNV5.js';
import './chunks/chunk-H4SJAY3E.js';
import './chunks/chunk-NRSSXRVS.js';
import './chunks/chunk-QKK3NKYN.js';
import './chunks/chunk-VTSXHYMT.js';
import './chunks/chunk-FNH5S7FD.js';
import './chunks/chunk-N7ILVN7U.js';
import './chunks/chunk-GE45ZFMY.js';
import './chunks/chunk-7FVMKTUJ.js';
import './chunks/chunk-4WEPWNB2.js';
import './chunks/chunk-NVFMRJVN.js';
import './chunks/chunk-HZSUQI6T.js';
import './chunks/chunk-MHHAGIRI.js';
import './chunks/chunk-4EBMB2HT.js';
import './chunks/chunk-YKIOTB2J.js';
import './chunks/chunk-IVWVADOA.js';
import './chunks/chunk-FGT5HSCH.js';
import './chunks/chunk-UBC7ZJMJ.js';
import './chunks/chunk-QYSGNYZ4.js';
import './chunks/chunk-2DGYZ3AT.js';
import './chunks/chunk-U6WW63HQ.js';
import './chunks/chunk-KMLSRSZ2.js';
import './chunks/chunk-YR2WIJFK.js';
import './chunks/chunk-OBC7ZDNY.js';
import './chunks/chunk-DQEIJ6AA.js';
import './chunks/chunk-HVROBH6S.js';
import './chunks/chunk-4YWO2LEM.js';
import './chunks/chunk-S7OQOFKY.js';
import './chunks/chunk-3A3NMPK7.js';
import './chunks/chunk-KAWCVYT7.js';
import './chunks/chunk-Y2E5LSRA.js';
import './chunks/chunk-MH2CV2WS.js';
import './chunks/chunk-UXSS3YXT.js';
import './chunks/chunk-A6X7WAFA.js';
import './chunks/chunk-56GYOYH6.js';
import './chunks/chunk-EZNG2RX5.js';
import './chunks/chunk-4XZ5VLZQ.js';
import './chunks/chunk-OBWYPIXU.js';
import './chunks/chunk-IGIIT6ND.js';
import './chunks/chunk-APKNUUBH.js';
import './chunks/chunk-W47WM6LO.js';
import './chunks/chunk-LINOWMLN.js';
import './chunks/chunk-26ZQYIUZ.js';
import './chunks/chunk-BUNPCUTG.js';
import './chunks/chunk-L6L3JFBG.js';
import './chunks/chunk-WQYSIU7X.js';
import './chunks/chunk-KGIJMXAP.js';
import './chunks/chunk-LBXYQH7K.js';
import './chunks/chunk-DT22F76Z.js';
import './chunks/chunk-OTPQATWR.js';
import './chunks/chunk-6SVCWHLM.js';
import './chunks/chunk-FJ2OEY6P.js';
import './chunks/chunk-3XSNY5HY.js';
import './chunks/chunk-YFCPYTCS.js';
import './chunks/chunk-M2YWWI52.js';
import './chunks/chunk-TCFNQMIV.js';
import './chunks/chunk-ZGFCCRQO.js';
import './chunks/chunk-ZB2Y4MES.js';
import './chunks/chunk-FPO3FMZ7.js';
import './chunks/chunk-HZPJZWMN.js';
import './chunks/chunk-4OVZOH72.js';
import './chunks/chunk-DJG366PE.js';
import './chunks/chunk-NCWLE3EN.js';
import './chunks/chunk-5CFDMIAZ.js';
import './chunks/chunk-PTJHYVQF.js';
import './chunks/chunk-6WF5PHO2.js';
import './chunks/chunk-7JCXTZWI.js';
import './chunks/chunk-RBPUEIYY.js';
import './chunks/chunk-A7U6TVNK.js';
import './chunks/chunk-AJFBO2QS.js';
import './chunks/chunk-X3SG235Y.js';
import './chunks/chunk-A3KOXRAG.js';
import './chunks/chunk-VE4U5GVE.js';
import './chunks/chunk-UMWH4QNT.js';
import './chunks/chunk-VPSTUDLO.js';
import { isString, listenEvent, DOMEvent } from 'maverick.js/std';
export { appendTriggerEvent, findTriggerEvent, hasTriggerEvent, walkTriggerEventChain } from 'maverick.js/std';

var LibASSTextRenderer = class {
  constructor(loader, config) {
    this.loader = loader;
    this.config = config;
    this.priority = 1;
    this.i = null;
    this.m = null;
    this._d = /(ssa|ass)$/;
  }
  canRender(track) {
    return !!track.src && (isString(track.type) && this._d.test(track.type) || this._d.test(track.src));
  }
  attach(video) {
    this.loader().then(async (mod) => {
      this.i = new mod.default({
        ...this.config,
        video,
        subUrl: this.m?.src || ""
      });
      listenEvent(this.i, "ready", () => {
        const canvas = this.i?.xi;
        if (canvas)
          canvas.style.pointerEvents = "none";
      });
      listenEvent(this.i, "error", (event) => {
        if (this.m) {
          this.m[TEXT_TRACK_READY_STATE] = 3;
          this.m.dispatchEvent(
            new DOMEvent("error", {
              trigger: event,
              detail: event.error
            })
          );
        }
      });
    });
  }
  changeTrack(track) {
    if (!track || track.readyState === 3) {
      this.$d();
    } else if (this.m !== track) {
      this.i?.setTrackByUrl(track.src);
      this.m = track;
    }
  }
  detach() {
    this.$d();
  }
  $d() {
    this.i?.freeTrack();
    this.m = null;
  }
};

export { LibASSTextRenderer };
