// vidstack@0.6.9/dist/server/chunks/chunk-A4DBFVNS.js
import { round } from './chunk-N2X5VJTG.js';
import { renderRadioGroup } from './chunk-TE367BQX.js';
import { ClassManager } from './chunk-XCUHASYJ.js';
import { RadioGroup } from './chunk-5RWMH5KW.js';
import { Radio } from './chunk-SKUE5DXF.js';
import { MenuItems } from './chunk-3QXOFPLA.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_ssr } from 'maverick.js/ssr';
import { computed, effect, peek } from 'maverick.js';
import { defineElement } from 'maverick.js/element';

var $$_templ = ["<!$><span><!$>", "</span>"];
var $$_templ_2 = ['<!$><span part="label">', "</span>"];
var $$_templ_3 = ['<!$><span part="info">', "</span>"];
var QualityMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this._sortedQualities = computed(() => {
      const { qualities } = this._media.$store;
      return [...qualities()].sort(
        (a, b) => b.height === a.height ? b.bitrate - a.bitrate : b.height - a.height
      );
    });
    this._media = useMedia();
  }
  onConnect(el) {
    effect(this._watchControllerDisabled.bind(this));
    effect(this._watchHintText.bind(this));
    const { radioClass, radioCheckClass } = this.$props;
    new ClassManager(el)._observe("media-radio", radioClass)._observe('[part="check"]', radioCheckClass);
  }
  _watchHintText() {
    const { autoLabel } = this.$props, { autoQuality, quality } = this._media.$store, qualityText = quality() ? quality().height + "p" : "";
    this._menu._hint.set(!autoQuality() ? qualityText : autoLabel() + ` (${qualityText})`);
  }
  _watchControllerDisabled() {
    const { qualities } = this._media.$store;
    this._menu._disable(qualities().length === 0);
  }
  _isDisabled() {
    const { canSetQuality, qualities } = this._media.$store;
    return !canSetQuality() || qualities().length === 0;
  }
  _onChange(event) {
    if (this._isDisabled())
      return;
    const radioGroup = event.target, value = radioGroup.value;
    if (value === "auto") {
      this._media.remote.changeQuality(-1, event);
      return;
    }
    const { qualities } = this._media.$store, index = peek(qualities).findIndex((quality) => this._getQualityId(quality) === value);
    if (index >= 0)
      this._media.remote.changeQuality(index, event);
  }
  _getValue() {
    const { quality, autoQuality } = this._media.$store;
    if (autoQuality())
      return "auto";
    const currentQuality = quality();
    return currentQuality ? this._getQualityId(currentQuality) : "auto";
  }
  _getQualityId(quality) {
    return quality.height + "_" + quality.bitrate;
  }
  _getOptions() {
    const { autoLabel, hideBitrate } = this.$props;
    return [
      { value: "auto", content: () => $$_ssr($$_templ, autoLabel) },
      ...this._sortedQualities().map((quality) => {
        const rate = `${round(quality.bitrate / 1e6, 2)} Mbps`;
        return {
          value: this._getQualityId(quality),
          content: () => [$$_ssr($$_templ_2, quality.height + "p"), !hideBitrate() && $$_ssr($$_templ_3, rate)]
        };
      })
    ];
  }
  render() {
    const { radioGroupClass } = this.$props;
    return renderRadioGroup({
      value: this._getValue.bind(this),
      options: this._getOptions.bind(this),
      radioGroupClass,
      onChange: this._onChange.bind(this)
    });
  }
};
QualityMenuItems.el = defineElement({
  tagName: "media-quality-menu-items",
  props: {
    autoLabel: "Auto",
    hideBitrate: false,
    radioGroupClass: null,
    radioClass: null,
    radioCheckClass: null
  }
});
QualityMenuItems.register = [RadioGroup, Radio];

export { QualityMenuItems };
