// vidstack@0.6.9/dist/server/chunks/chunk-SKUE5DXF.js
import { radioGroupContext } from './chunk-4L7R5OAB.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { onPress } from './chunk-JTTKMFDW.js';
import { menuContext } from './chunk-HSL7QSWM.js';
import { FocusVisibleController } from './chunk-RLGAJ5QB.js';
import { __decorateClass } from './chunk-VPSTUDLO.js';
import { $$_ssr } from 'maverick.js/ssr';
import { signal, effect, hasProvidedContext, useContext, peek } from 'maverick.js';
import { defineElement, prop, Component } from 'maverick.js/element';

var $$_templ = ['<!$><div part="check"></div>'];
var Radio = class extends Component {
  constructor(instance) {
    super(instance);
    this._checked = signal(false);
    this._controller = {
      _value: this.$props.value,
      _check: this._check.bind(this),
      _onCheck: null
    };
    new FocusVisibleController(instance);
  }
  get checked() {
    return this._checked();
  }
  onAttach() {
    effect(this._watchValue.bind(this));
    this._addToGroup();
    const isMenuItem = hasProvidedContext(menuContext);
    this.setAttributes({
      value: this.$props.value,
      role: isMenuItem ? "menuitemradio" : "radio",
      tabindex: isMenuItem ? -1 : 0,
      "aria-checked": $ariaBool(this._checked)
    });
  }
  onConnect(el) {
    this._addToGroup();
    onPress(el, this._onPress.bind(this));
  }
  onDisconnect() {
    const group = useContext(radioGroupContext);
    group.remove(this._controller);
  }
  _addToGroup() {
    const group = useContext(radioGroupContext);
    group.add(this._controller);
  }
  _watchValue() {
    var _a, _b;
    const { value } = this.$props, newValue = value();
    if (peek(this._checked)) {
      (_b = (_a = this._controller)._onCheck) == null ? void 0 : _b.call(_a, newValue);
    }
  }
  _onPress(event) {
    var _a, _b;
    if (peek(this._checked))
      return;
    this._checked.set(true);
    this.dispatch("change", { trigger: event });
    (_b = (_a = this._controller)._onCheck) == null ? void 0 : _b.call(_a, peek(this.$props.value), event);
  }
  _check(value, trigger) {
    if (peek(this._checked) === value)
      return;
    this._checked.set(value);
    this.dispatch("change", { trigger });
  }
  render() {
    return $$_ssr($$_templ);
  }
};
Radio.el = defineElement({
  tagName: "media-radio",
  props: { value: "" }
});
__decorateClass([
  prop
], Radio.prototype, "checked", 1);

export { Radio };
