// vidstack@0.6.9/dist/server/chunks/chunk-4F5G36NV.js
import { renderRadioGroup } from './chunk-TE367BQX.js';
import { ClassManager } from './chunk-XCUHASYJ.js';
import { RadioGroup } from './chunk-5RWMH5KW.js';
import { Radio } from './chunk-SKUE5DXF.js';
import { MenuItems } from './chunk-3QXOFPLA.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_ssr } from 'maverick.js/ssr';
import { effect } from 'maverick.js';
import { defineElement } from 'maverick.js/element';

var $$_templ = ['<!$><span part="label"><!$>', "</span>"];
var AudioMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this._media = useMedia();
  }
  onConnect(el) {
    effect(this._watchControllerDisabled.bind(this));
    effect(this._watchHintText.bind(this));
    const { radioClass, radioCheckClass } = this.$props;
    new ClassManager(el)._observe("media-radio", radioClass)._observe('[part="check"]', radioCheckClass);
  }
  _watchHintText() {
    const { emptyLabel } = this.$props, { audioTrack } = this._media.$store, track = audioTrack();
    this._menu._hint.set((track == null ? void 0 : track.label) ?? emptyLabel());
  }
  _watchControllerDisabled() {
    this._menu._disable(this._isDisabled());
  }
  _isDisabled() {
    const { audioTracks } = this._media.$store;
    return audioTracks().length === 0;
  }
  _onChange(event) {
    if (this._isDisabled())
      return;
    const radioGroup = event.target;
    const value = radioGroup.value, index = this._media.audioTracks.toArray().findIndex((track) => track.label.toLowerCase() === value);
    if (index >= 0)
      this._media.remote.changeAudioTrack(index, event);
  }
  _getValue() {
    const { audioTrack } = this._media.$store;
    const track = audioTrack();
    return track ? track.label.toLowerCase() : "";
  }
  _getOptions() {
    const { audioTracks } = this._media.$store;
    return audioTracks().map((track) => ({
      value: track.label.toLowerCase(),
      content: () => $$_ssr($$_templ, track.label)
    }));
  }
  render() {
    const { radioGroupClass } = this.$props;
    return renderRadioGroup({
      value: this._getValue.bind(this),
      options: this._getOptions.bind(this),
      radioGroupClass,
      onChange: this._onChange.bind(this)
    });
  }
};
AudioMenuItems.el = defineElement({
  tagName: "media-audio-menu-items",
  props: {
    emptyLabel: "Default",
    radioClass: null,
    radioGroupClass: null,
    radioCheckClass: null
  }
});
AudioMenuItems.register = [RadioGroup, Radio];

export { AudioMenuItems };
