// vidstack@0.6.9/dist/server/chunks/chunk-4BWBKFM6.js
import { Thumbnail } from './chunk-MCANHAYI.js';
import { formatTime, formatSpokenTime } from './chunk-JY53PGKC.js';
import { round } from './chunk-N2X5VJTG.js';
import { renderRadioGroup } from './chunk-TE367BQX.js';
import { ClassManager } from './chunk-XCUHASYJ.js';
import { RadioGroup } from './chunk-5RWMH5KW.js';
import { Radio } from './chunk-SKUE5DXF.js';
import { MenuItems } from './chunk-3QXOFPLA.js';
import { isCueActive, onTrackChapterChange } from './chunk-NOPW2CG3.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { $$_ssr, $$_custom_element } from 'maverick.js/ssr';
import { signal, effect, peek } from 'maverick.js';
import { defineElement } from 'maverick.js/element';
import { isKeyboardEvent, listenEvent, setStyle, isNumber } from 'maverick.js/std';

var $$_templ = ["<!$><media-thumbnail", "", "</media-thumbnail>"];
var $$_templ_2 = ['<!$><div part="content"><div part="title">', '</div><div part="start-time">', '</div><div part="duration">', "</div></div>"];
var ChaptersMenuItems = class extends MenuItems {
  constructor(instance) {
    super(instance);
    this._index = signal(0);
    this._track = signal(null);
    this._media = useMedia();
  }
  onAttach(el) {
    super.onAttach(el);
    this._menu._attachObserver({
      _onOpen: this._onOpen.bind(this)
    });
    this.setAttributes({
      "data-thumbnails": this._hasThumbnails.bind(this)
    });
  }
  _onOpen(trigger) {
    if (isKeyboardEvent(trigger))
      return;
    requestAnimationFrame(() => {
      const checked = this.el.querySelector('media-radio[aria-checked="true"]');
      checked.scrollIntoView({ block: "center" });
    });
  }
  onConnect(el) {
    effect(this._watchCurrentTime.bind(this));
    effect(this._watchControllerDisabled.bind(this));
    this._onTrackModeChange();
    listenEvent(this._media.textTracks, "mode-change", this._onTrackModeChange.bind(this));
    const {
      chapterClass,
      thumbnailClass,
      contentClass,
      titleClass,
      startTimeClass,
      durationClass
    } = this.$props;
    new ClassManager(el)._observe('[part="chapter"]', chapterClass)._observe('[part="thumbnail"]', thumbnailClass)._observe('[part="content"]', contentClass)._observe('[part="title"]', titleClass)._observe('[part="start-time"]', startTimeClass)._observe('[part="duration"]', durationClass);
  }
  _hasThumbnails() {
    const { thumbnailCues } = this._media.$store;
    return thumbnailCues().length > 0;
  }
  _watchCurrentTime() {
    if (!this._menu._expanded())
      return;
    const track = this._track(), { currentTime } = this._media.$store;
    if (!track) {
      this._index.set(-1);
      return;
    }
    const time = currentTime(), activeCueIndex = track.cues.findIndex((cue) => isCueActive(cue, time));
    this._index.set(activeCueIndex);
    if (activeCueIndex >= 0) {
      const cue = track.cues[activeCueIndex], radio = this.el.querySelector(`shadow-root media-radio[aria-checked='true']`), playedPercent = (time - cue.startTime) / (cue.endTime - cue.startTime) * 100;
      radio && setStyle(radio, "--played-percent", round(playedPercent, 3) + "%");
    }
  }
  _watchControllerDisabled() {
    this._menu._disable(this._isDisabled());
  }
  _isDisabled() {
    const track = this._track();
    return !track || !track.cues.length;
  }
  _onChange(event) {
    var _a;
    if (this._isDisabled() || !event.trigger)
      return;
    const index = +event.target.value, cues = (_a = this._track()) == null ? void 0 : _a.cues;
    if (isNumber(index) && (cues == null ? void 0 : cues[index])) {
      this._index.set(index);
      this._media.remote.seek(cues[index].startTime, event);
    }
  }
  _onTrackModeChange() {
    onTrackChapterChange(this._media.textTracks, peek(this._track), this._track.set);
  }
  _getValue() {
    return this._index() + "";
  }
  _getOptions() {
    const track = this._track();
    if (!track)
      return [];
    return track.cues.map((cue, i) => ({
      value: i + "",
      content: () => [
        this._hasThumbnails() && $$_ssr($$_templ, ...$$_custom_element("media-thumbnail", { time: cue.startTime }, [{ "part": "thumbnail" }])),
        $$_ssr(
          $$_templ_2,
          cue.text,
          formatTime(cue.startTime, false, cue.startTime >= 3600),
          formatSpokenTime(cue.endTime - cue.startTime)
        )
      ]
    }));
  }
  render() {
    const { containerClass } = this.$props;
    return renderRadioGroup({
      part: "chapter",
      value: this._getValue.bind(this),
      options: this._getOptions.bind(this),
      radioGroupClass: containerClass,
      onChange: this._onChange.bind(this)
    });
  }
};
ChaptersMenuItems.el = defineElement({
  tagName: "media-chapters-menu-items",
  props: {
    containerClass: null,
    chapterClass: null,
    thumbnailClass: null,
    contentClass: null,
    titleClass: null,
    startTimeClass: null,
    durationClass: null
  }
});
ChaptersMenuItems.register = [Thumbnail, RadioGroup, Radio];

export { ChaptersMenuItems };
