// vidstack@0.6.9/dist/server/chunks/chunk-MY2FGNE2.js
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { setAttributeIfEmpty, setARIALabel, onPress } from './chunk-JTTKMFDW.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-HPMGEKCS.js';
import { TooltipController } from './chunk-YQQPXCW6.js';
import { seek_backward_default } from './chunk-5XOURGNQ.js';
import { seek_forward_default } from './chunk-MUUSA6L7.js';
import { FocusVisibleController } from './chunk-RLGAJ5QB.js';
import { $$_create_component } from 'maverick.js/ssr';
import { defineElement, Component } from 'maverick.js/element';

var SeekButton = class extends Component {
  constructor(instance) {
    super(instance);
    this._media = useMedia();
    new FocusVisibleController(instance);
    new TooltipController(instance);
  }
  onAttach(el) {
    setAttributeIfEmpty(el, "tabindex", "0");
    setAttributeIfEmpty(el, "role", "button");
    setARIALabel(el, this._getLabel.bind(this));
    const { seconds, defaultAppearance } = this.$props;
    this.setAttributes({
      seconds,
      "default-appearance": defaultAppearance,
      "aria-hidden": $ariaBool(this._isHidden.bind(this)),
      "data-media-button": true
    });
  }
  onConnect(el) {
    onPress(el, this._onPress.bind(this));
  }
  _isHidden() {
    const { canSeek } = this._media.$store;
    return !canSeek();
  }
  _getLabel() {
    const { seconds } = this.$props;
    return `Seek ${seconds() > 0 ? "forward" : "backward"} ${seconds()} seconds`;
  }
  _onPress(event) {
    const { seconds, disabled } = this.$props;
    if (disabled())
      return;
    const { currentTime } = this._media.$store, seekTo = currentTime() + seconds();
    this._media.remote.seek(seekTo, event);
  }
  render() {
    return [
      $$_create_component(Icon, { paths: seek_backward_default, slot: "backward" }),
      $$_create_component(Icon, { paths: seek_forward_default, slot: "forward" })
    ];
  }
};
SeekButton.el = defineElement({
  tagName: "media-seek-button",
  props: {
    disabled: false,
    defaultAppearance: false,
    seconds: 30
  }
});

export { SeekButton };
