// vidstack@0.6.9/dist/server/chunks/chunk-MBBDB3JG.js
import { toggleButtonProps, ToggleButton } from './chunk-RGSXXUQI.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { setARIALabel } from './chunk-JTTKMFDW.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-HPMGEKCS.js';
import { fullscreen_exit_default } from './chunk-E5QSZH7H.js';
import { fullscreen_default } from './chunk-MRMVJNCY.js';
import { $$_create_component } from 'maverick.js/ssr';
import { defineElement } from 'maverick.js/element';

var FullscreenButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this._keyShortcut = "toggleFullscreen";
  }
  onAttach(el) {
    this._media = useMedia();
    this._pressed = this._isPressed.bind(this);
    super.onAttach(el);
    setARIALabel(el, this._getLabel.bind(this));
    const { fullscreen } = this._media.$store;
    this.setAttributes({
      "aria-hidden": $ariaBool(this._isHidden.bind(this)),
      "data-fullscreen": fullscreen
    });
  }
  _onPress(event) {
    const remote = this._media.remote, target = this.$props.target();
    this._pressed() ? remote.exitFullscreen(target, event) : remote.enterFullscreen(target, event);
  }
  _isPressed() {
    const { fullscreen } = this._media.$store;
    return fullscreen();
  }
  _isHidden() {
    const { canFullscreen } = this._media.$store;
    return !canFullscreen();
  }
  _getLabel() {
    const { fullscreen } = this._media.$store;
    return fullscreen() ? "Exit Fullscreen" : "Enter Fullscreen";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: fullscreen_default, slot: "enter" }),
      $$_create_component(Icon, { paths: fullscreen_exit_default, slot: "exit" })
    ];
  }
};
FullscreenButton.el = defineElement({
  tagName: "media-fullscreen-button",
  props: {
    ...toggleButtonProps,
    target: "prefer-media"
  }
});

export { FullscreenButton };
