// vidstack@0.6.9/dist/server/chunks/chunk-ZZBQCJ7P.js
import { isTrackCaptionKind } from './chunk-ZG64VF3W.js';
import { toggleButtonProps, ToggleButton } from './chunk-RGSXXUQI.js';
import { $ariaBool } from './chunk-EVP5E6AO.js';
import { setARIALabel } from './chunk-JTTKMFDW.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-HPMGEKCS.js';
import { closed_captions_on_default } from './chunk-BIGBL5GY.js';
import { closed_captions_default } from './chunk-WAHVE4HI.js';
import { $$_create_component } from 'maverick.js/ssr';
import { defineElement } from 'maverick.js/element';

var CaptionButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this._keyShortcut = "toggleCaptions";
  }
  onAttach(el) {
    this._media = useMedia();
    this._pressed = this._isPressed.bind(this);
    super.onAttach(el);
    setARIALabel(el, this._getLabel.bind(this));
    this.setAttributes({
      "aria-hidden": $ariaBool(this._isHidden.bind(this))
    });
  }
  _onPress(event) {
    this._media.remote.toggleCaptions(event);
  }
  _isPressed() {
    const { textTrack } = this._media.$store, track = textTrack();
    return !!track && isTrackCaptionKind(track);
  }
  _isHidden() {
    const { textTracks } = this._media.$store;
    return textTracks().filter(isTrackCaptionKind).length == 0;
  }
  _getLabel() {
    const { textTrack } = this._media.$store;
    return textTrack() ? "Closed-Captions Off" : "Closed-Captions On";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: closed_captions_on_default, slot: "on" }),
      $$_create_component(Icon, { paths: closed_captions_default, slot: "off" })
    ];
  }
};
CaptionButton.el = defineElement({
  tagName: "media-caption-button",
  props: toggleButtonProps
});

export { CaptionButton };
