// vidstack/dist/server/chunks/chunk-3MX4LIKX.js
import { sliderValueFormatContext } from './chunk-U43VNZZW.js';
import { SliderStoreFactory } from './chunk-53B74NMI.js';
import { formatTime } from './chunk-JY53PGKC.js';
import { round } from './chunk-N2X5VJTG.js';
import { $$_ssr } from 'maverick.js/ssr';
import { useStore, useContext, computed } from 'maverick.js';
import { defineElement, Component } from 'maverick.js/element';

var $$_templ = ["<!$><span><!$>", "</span>"];
var SliderValue = class extends Component {
  onAttach() {
    this._slider = useStore(SliderStoreFactory);
    this._format = useContext(sliderValueFormatContext);
    this._text = computed(this._getText.bind(this));
  }
  _getText() {
    var _a, _b;
    const { type, format, decimalPlaces, padHours, padMinutes, showHours } = this.$props;
    const { value: sliderValue, pointerValue, min, max } = this._slider;
    const value = type() === "current" ? sliderValue() : pointerValue();
    if (format() === "percent") {
      const range = max() - min();
      const percent = value / range * 100;
      return (this._format.percent ?? round)(percent, decimalPlaces()) + "\uFE6A";
    } else if (format() === "time") {
      return (this._format.time ?? formatTime)(value, padHours(), padMinutes(), showHours());
    } else {
      return ((_b = (_a = this._format).value) == null ? void 0 : _b.call(_a, value)) ?? value.toFixed(2);
    }
  }
  render() {
    return $$_ssr($$_templ, this._text);
  }
};
SliderValue.el = defineElement({
  tagName: "media-slider-value",
  props: {
    type: "current",
    format: void 0,
    showHours: false,
    padHours: false,
    padMinutes: false,
    decimalPlaces: 2
  }
});

export { SliderValue };
