// vidstack/dist/server/chunks/chunk-BQKGS4S5.js
import { toggleButtonProps, ToggleButton } from './chunk-RGSXXUQI.js';
import { setARIALabel } from './chunk-JTTKMFDW.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { Icon } from './chunk-HPMGEKCS.js';
import { volume_high_default } from './chunk-64OI64HO.js';
import { volume_low_default } from './chunk-VE2HQ3IX.js';
import { mute_default } from './chunk-HZUXRM2O.js';
import { $$_create_component } from 'maverick.js/ssr';
import { defineElement } from 'maverick.js/element';

var MuteButton = class extends ToggleButton {
  constructor() {
    super(...arguments);
    this._keyShortcut = "toggleMuted";
  }
  onAttach(el) {
    this._media = useMedia();
    this._pressed = this._isPressed.bind(this);
    setARIALabel(el, this._getLabel.bind(this));
    this.setAttributes({
      "data-muted": this._pressed,
      "data-volume": this._getVolumeText.bind(this)
    });
    super.onAttach(el);
  }
  _onPress(event) {
    const remote = this._media.remote;
    this._pressed() ? remote.unmute(event) : remote.mute(event);
  }
  _isPressed() {
    const { muted, volume } = this._media.$store;
    return muted() || volume() === 0;
  }
  _getLabel() {
    return this._pressed() ? "Unmute" : "Mute";
  }
  _getVolumeText() {
    const { muted, volume } = this._media.$store, $volume = volume();
    if (muted() || $volume === 0)
      return "muted";
    else if ($volume >= 0.5)
      return "high";
    else if ($volume < 0.5)
      return "low";
  }
  render() {
    return [
      $$_create_component(Icon, { paths: volume_high_default, slot: "volume-high" }),
      $$_create_component(Icon, { paths: volume_low_default, slot: "volume-low" }),
      $$_create_component(Icon, { paths: mute_default, slot: "volume-muted" })
    ];
  }
};
MuteButton.el = defineElement({
  tagName: "media-mute-button",
  props: toggleButtonProps
});

export { MuteButton };
