// vidstack/dist/server/chunks/chunk-J3NSFYPM.js
import { setARIALabel, onPress } from './chunk-JTTKMFDW.js';
import { useMedia } from './chunk-ZIBTY3GT.js';
import { FocusVisibleController } from './chunk-RLGAJ5QB.js';
import { $$_ssr } from 'maverick.js/ssr';
import { defineElement, Component } from 'maverick.js/element';

var $$_templ = ['<!$><div part="container"><div part="text">LIVE</div></div>'];
var LiveIndicator = class extends Component {
  constructor(instance) {
    super(instance);
    this._media = useMedia();
    new FocusVisibleController(instance);
  }
  onAttach(el) {
    const { live, liveEdge } = this._media.$store;
    setARIALabel(el, this._getLabel.bind(this));
    this.setAttributes({
      tabindex: this._getTabIndex.bind(this),
      role: this._getRole.bind(this),
      "data-live": live,
      "data-live-edge": liveEdge,
      "data-media-button": true
    });
  }
  onConnect(el) {
    onPress(el, this._onPress.bind(this));
  }
  _getLabel() {
    const { live } = this._media.$store;
    return live() ? "Go live" : null;
  }
  _getTabIndex() {
    const { live } = this._media.$store;
    return live() ? 0 : null;
  }
  _getRole() {
    const { live } = this._media.$store;
    return live() ? "button" : null;
  }
  _onPress(event) {
    const { liveEdge } = this._media.$store;
    if (liveEdge())
      return;
    this._media.remote.seekToLiveEdge(event);
  }
  render() {
    return $$_ssr($$_templ);
  }
};
LiveIndicator.el = defineElement({
  tagName: "media-live-indicator"
});

export { LiveIndicator };
