// vidstack@0.6.9/dist/dev/index.js
export { CommunitySkin } from './chunks/chunk-37JVNHU6.js';
export { Tooltip } from './chunks/chunk-DON7JMVQ.js';
export { VolumeSlider } from './chunks/chunk-JBX3YXBH.js';
export { SliderThumbnail } from './chunks/chunk-MJRWGNJB.js';
export { SliderValue } from './chunks/chunk-UKG5Z2SB.js';
export { SliderVideo } from './chunks/chunk-AWEYXSJQ.js';
export { TimeSlider } from './chunks/chunk-5PVPU33U.js';
export { Slider, sliderProps } from './chunks/chunk-LWTGALKS.js';
import './chunks/chunk-U43VNZZW.js';
export { SliderStoreFactory } from './chunks/chunk-53B74NMI.js';
export { Time } from './chunks/chunk-LZTNZZHJ.js';
export { PlaybackRateMenuItems } from './chunks/chunk-BLODLME6.js';
export { AudioTrackList, FullscreenController, List, MEDIA_KEY_SHORTCUTS, MediaRemoteControl, MediaStoreFactory, MediaUserController, Player, ScreenOrientationController, TextRenderers, TextTrackList, TimeRange, VideoQualityList, canFullscreen, getTimeRangesEnd, getTimeRangesStart, isAudioProvider, isHLSProvider, isHTMLAudioElement, isHTMLMediaElement, isHTMLVideoElement, isVideoProvider, softResetMediaStore } from './chunks/chunk-VI4ON6IC.js';
import './chunks/chunk-LWO4CMSF.js';
import './chunks/chunk-HL25N2T5.js';
import './chunks/chunk-QZI2LQVA.js';
import './chunks/chunk-ANZD5B3N.js';
import './chunks/chunk-JELAJF2G.js';
export { Poster } from './chunks/chunk-VTFMYSE2.js';
export { QualityMenuButton } from './chunks/chunk-SJD42WNH.js';
export { QualityMenuItems } from './chunks/chunk-57KDYM5F.js';
export { SeekButton } from './chunks/chunk-QTDX7HDE.js';
export { Menu } from './chunks/chunk-2NMX37TU.js';
export { MuteButton } from './chunks/chunk-XBG5E767.js';
export { Outlet } from './chunks/chunk-FY36WF27.js';
import './chunks/chunk-FRNAFKFA.js';
import './chunks/chunk-WSYKLWQE.js';
export { PIPButton } from './chunks/chunk-2DWMEI5Q.js';
export { PlayButton } from './chunks/chunk-QVMUDSL4.js';
export { PlaybackRateMenuButton } from './chunks/chunk-UOQZCGPO.js';
export { CaptionsMenuItems } from './chunks/chunk-E5ZWM3RR.js';
export { Captions } from './chunks/chunk-L4H3J6NX.js';
export { ChaptersMenuItems } from './chunks/chunk-EYRSCQBE.js';
export { Thumbnail } from './chunks/chunk-AJO5IWRQ.js';
import './chunks/chunk-JY53PGKC.js';
import './chunks/chunk-CVLY5S52.js';
export { FullscreenButton } from './chunks/chunk-V32T6ASS.js';
export { Gesture } from './chunks/chunk-WIWKZOJA.js';
export { LiveIndicator } from './chunks/chunk-LTLQUHKO.js';
export { AudioMenuButton } from './chunks/chunk-MQYRFZNM.js';
export { AudioMenuItems } from './chunks/chunk-ATST7NI2.js';
import './chunks/chunk-KRJ7Q224.js';
import './chunks/chunk-XCUHASYJ.js';
export { RadioGroup } from './chunks/chunk-MIWMXUN7.js';
export { Radio } from './chunks/chunk-P2N5WSLT.js';
import './chunks/chunk-4L7R5OAB.js';
export { MenuItems } from './chunks/chunk-6F6XCALD.js';
export { BufferingIndicator } from './chunks/chunk-LFUONOET.js';
export { CaptionButton } from './chunks/chunk-JCI4LGRO.js';
import { TEXT_TRACK_READY_STATE } from './chunks/chunk-2SNW2HZF.js';
export { TextTrack, isTrackCaptionKind } from './chunks/chunk-2SNW2HZF.js';
import './chunks/chunk-IVW6GMZV.js';
import './chunks/chunk-NOPW2CG3.js';
export { ToggleButton, toggleButtonProps } from './chunks/chunk-252ZP55V.js';
import './chunks/chunk-EVP5E6AO.js';
import './chunks/chunk-HQQTSFXD.js';
export { mediaContext } from './chunks/chunk-ZIBTY3GT.js';
export { CaptionsMenuButton } from './chunks/chunk-Y76E56TL.js';
import './chunks/chunk-HAOWXMSO.js';
export { MenuButton } from './chunks/chunk-MX7CJR52.js';
import './chunks/chunk-HSL7QSWM.js';
import './chunks/chunk-OSS74BPE.js';
import './chunks/chunk-YQQPXCW6.js';
import './chunks/chunk-RBTESKWW.js';
import './chunks/chunk-R6FSMW74.js';
import './chunks/chunk-MF43X3Y3.js';
import './chunks/chunk-LBH32PRG.js';
import './chunks/chunk-KQUCP2EG.js';
import './chunks/chunk-DZHDFPQR.js';
import './chunks/chunk-Q2XL2SHG.js';
import './chunks/chunk-Z3R7IJO3.js';
import './chunks/chunk-2IK7NGIX.js';
import './chunks/chunk-KITPHWQZ.js';
import './chunks/chunk-HKZBS6I2.js';
import './chunks/chunk-WPVULLF5.js';
import './chunks/chunk-WNZBENRE.js';
import './chunks/chunk-TYOX6JVW.js';
import './chunks/chunk-ZUP6VWH2.js';
import './chunks/chunk-TT5YQ2JU.js';
import './chunks/chunk-3V72BDJY.js';
import './chunks/chunk-764WJ7YC.js';
import './chunks/chunk-J24QS7SA.js';
import './chunks/chunk-5AX42AEN.js';
import './chunks/chunk-I5ZNHUW2.js';
import './chunks/chunk-W4QNKYWE.js';
import './chunks/chunk-YGXIG4UH.js';
import './chunks/chunk-DJRMQYAX.js';
import './chunks/chunk-KU5TXDFG.js';
import './chunks/chunk-PDX6CZZG.js';
import './chunks/chunk-B37QG5XE.js';
import './chunks/chunk-RKG5SB2X.js';
import './chunks/chunk-IDM7ODL5.js';
import './chunks/chunk-CPWHUQBL.js';
import './chunks/chunk-EOTLQ6GG.js';
import './chunks/chunk-BQ2XLH4F.js';
import './chunks/chunk-E23QUJXR.js';
import './chunks/chunk-BM7OTNV5.js';
import './chunks/chunk-H4SJAY3E.js';
import './chunks/chunk-NRSSXRVS.js';
import './chunks/chunk-QKK3NKYN.js';
import './chunks/chunk-VTSXHYMT.js';
import './chunks/chunk-FNH5S7FD.js';
import './chunks/chunk-N7ILVN7U.js';
import './chunks/chunk-GE45ZFMY.js';
import './chunks/chunk-7FVMKTUJ.js';
import './chunks/chunk-4WEPWNB2.js';
import './chunks/chunk-NVFMRJVN.js';
import './chunks/chunk-HZSUQI6T.js';
import './chunks/chunk-MHHAGIRI.js';
import './chunks/chunk-4EBMB2HT.js';
import './chunks/chunk-YKIOTB2J.js';
import './chunks/chunk-IVWVADOA.js';
import './chunks/chunk-FGT5HSCH.js';
import './chunks/chunk-UBC7ZJMJ.js';
import './chunks/chunk-QYSGNYZ4.js';
import './chunks/chunk-2DGYZ3AT.js';
import './chunks/chunk-U6WW63HQ.js';
import './chunks/chunk-KMLSRSZ2.js';
import './chunks/chunk-YR2WIJFK.js';
import './chunks/chunk-OBC7ZDNY.js';
import './chunks/chunk-DQEIJ6AA.js';
import './chunks/chunk-HVROBH6S.js';
import './chunks/chunk-4YWO2LEM.js';
import './chunks/chunk-S7OQOFKY.js';
import './chunks/chunk-3A3NMPK7.js';
import './chunks/chunk-KAWCVYT7.js';
import './chunks/chunk-Y2E5LSRA.js';
import './chunks/chunk-MH2CV2WS.js';
import './chunks/chunk-UXSS3YXT.js';
import './chunks/chunk-A6X7WAFA.js';
import './chunks/chunk-56GYOYH6.js';
import './chunks/chunk-EZNG2RX5.js';
import './chunks/chunk-4XZ5VLZQ.js';
import './chunks/chunk-OBWYPIXU.js';
import './chunks/chunk-IGIIT6ND.js';
import './chunks/chunk-APKNUUBH.js';
import './chunks/chunk-W47WM6LO.js';
import './chunks/chunk-LINOWMLN.js';
import './chunks/chunk-26ZQYIUZ.js';
import './chunks/chunk-BUNPCUTG.js';
import './chunks/chunk-L6L3JFBG.js';
import './chunks/chunk-WQYSIU7X.js';
import './chunks/chunk-KGIJMXAP.js';
import './chunks/chunk-LBXYQH7K.js';
import './chunks/chunk-DT22F76Z.js';
import './chunks/chunk-OTPQATWR.js';
import './chunks/chunk-6SVCWHLM.js';
import './chunks/chunk-FJ2OEY6P.js';
import './chunks/chunk-3XSNY5HY.js';
import './chunks/chunk-YFCPYTCS.js';
import './chunks/chunk-M2YWWI52.js';
import './chunks/chunk-TCFNQMIV.js';
import './chunks/chunk-ZGFCCRQO.js';
import './chunks/chunk-ZB2Y4MES.js';
import './chunks/chunk-FPO3FMZ7.js';
import './chunks/chunk-HZPJZWMN.js';
import './chunks/chunk-4OVZOH72.js';
import './chunks/chunk-DJG366PE.js';
import './chunks/chunk-NCWLE3EN.js';
import './chunks/chunk-5CFDMIAZ.js';
import './chunks/chunk-PTJHYVQF.js';
import './chunks/chunk-6WF5PHO2.js';
import './chunks/chunk-7JCXTZWI.js';
import './chunks/chunk-RBPUEIYY.js';
import './chunks/chunk-A7U6TVNK.js';
import './chunks/chunk-AJFBO2QS.js';
import './chunks/chunk-X3SG235Y.js';
import './chunks/chunk-A3KOXRAG.js';
import './chunks/chunk-VE4U5GVE.js';
import './chunks/chunk-IQ5YT2BQ.js';
import './chunks/chunk-VPSTUDLO.js';
import { isString, listenEvent, DOMEvent } from 'maverick.js/std';
export { appendTriggerEvent, findTriggerEvent, hasTriggerEvent, walkTriggerEventChain } from 'maverick.js/std';

var LibASSTextRenderer = class {
  constructor(loader, config) {
    this.loader = loader;
    this.config = config;
    this.priority = 1;
    this._instance = null;
    this._track = null;
    this._typeRE = /(ssa|ass)$/;
  }
  canRender(track) {
    return !!track.src && (isString(track.type) && this._typeRE.test(track.type) || this._typeRE.test(track.src));
  }
  attach(video) {
    this.loader().then(async (mod) => {
      this._instance = new mod.default({
        ...this.config,
        video,
        subUrl: this._track?.src || ""
      });
      listenEvent(this._instance, "ready", () => {
        const canvas = this._instance?._canvas;
        if (canvas)
          canvas.style.pointerEvents = "none";
      });
      listenEvent(this._instance, "error", (event) => {
        if (this._track) {
          this._track[TEXT_TRACK_READY_STATE] = 3;
          this._track.dispatchEvent(
            new DOMEvent("error", {
              trigger: event,
              detail: event.error
            })
          );
        }
      });
    });
  }
  changeTrack(track) {
    if (!track || track.readyState === 3) {
      this._freeTrack();
    } else if (this._track !== track) {
      this._instance?.setTrackByUrl(track.src);
      this._track = track;
    }
  }
  detach() {
    this._freeTrack();
  }
  _freeTrack() {
    this._instance?.freeTrack();
    this._track = null;
  }
};

// src/index.ts
{
  console.warn("[vidstack] dev mode!");
}

export { LibASSTextRenderer };
